In [2]:
import aegnn
import torch
import lightning.pytorch as pl

# import os
# os.environ["CUDA_LAUNCH_BLOCKING"] = "1"

In [3]:
import os
os.environ["AEGNN_DATA_DIR"] = "/home/ibrahim/event_graph_aegnn/test_codes/data"
print(os.environ["AEGNN_DATA_DIR"])

/home/ibrahim/event_graph_aegnn/test_codes/data


# Data visualization

In [4]:
data_module = aegnn.datasets.NCars(batch_size=16, shuffle=False, )


In [5]:
data_module

NCars[Train: None
Validation: None]

In [6]:
val_dataset = data_module.val_dataset

In [7]:
data_module.test_dataset = val_dataset

In [8]:
data_module = aegnn.datasets.NCars(batch_size=16, shuffle=False, )
data_module.setup()

UserWarning: No data found, check AEGNN_DATA_DIR environment variable!

In [ ]:
len(data_module.test_dataset)

8606

In [ ]:
data_module.train_dataset.files

['/home/ibrahim/event_graph_aegnn/test_codes/data/ncars/processed/training/sequence_1961',
 '/home/ibrahim/event_graph_aegnn/test_codes/data/ncars/processed/training/sequence_5556',
 '/home/ibrahim/event_graph_aegnn/test_codes/data/ncars/processed/training/sequence_9253',
 '/home/ibrahim/event_graph_aegnn/test_codes/data/ncars/processed/training/sequence_10095',
 '/home/ibrahim/event_graph_aegnn/test_codes/data/ncars/processed/training/sequence_9350',
 '/home/ibrahim/event_graph_aegnn/test_codes/data/ncars/processed/training/sequence_6639',
 '/home/ibrahim/event_graph_aegnn/test_codes/data/ncars/processed/training/sequence_10368',
 '/home/ibrahim/event_graph_aegnn/test_codes/data/ncars/processed/training/sequence_7203',
 '/home/ibrahim/event_graph_aegnn/test_codes/data/ncars/processed/training/sequence_762',
 '/home/ibrahim/event_graph_aegnn/test_codes/data/ncars/processed/training/sequence_6426',
 '/home/ibrahim/event_graph_aegnn/test_codes/data/ncars/processed/training/sequence_1168'

In [ ]:
import glob
import os


In [ ]:
data_module.root 

'/home/ibrahim/event_graph_aegnn/test_codes/data/ncars'

In [ ]:
g = data_module.train_dataset[0]

In [ ]:
g

Data(x=[5715, 4], pos=[5715, 3], file_id='sequence_1961', label=[1], y=[1], edge_index=[2, 151044], edge_attr=[151044, 3])

In [ ]:
g.x.shape

torch.Size([5715, 4])

In [ ]:
g.edge_index

tensor([[   0,    0,    0,  ..., 5714, 5714, 5714],
        [5234, 3871, 3563,  ..., 1153, 1321, 1339]])

In [ ]:
g.edge_attr

tensor([[0.5000, 0.5000, 0.5000],
        [0.5000, 0.5000, 0.5000],
        [0.5000, 0.5000, 0.5000],
        ...,
        [0.6667, 0.5000, 0.5000],
        [0.3333, 0.5000, 0.5000],
        [0.3333, 0.5000, 0.5000]])

In [ ]:
g.label

['background']

In [ ]:
g.y

tensor([1])

# Data pre-processing

 Convert event stream in event graph

## NCARS

In [ ]:
from aegnn.datasets.ncars import NCars

In [ ]:
ncars = NCars()

In [ ]:
ncars

NCars[Train: None
Validation: None]

In [ ]:
event_sample_file = ncars.raw_files("training")
event_sample_file

['/home/ibrahim/event_graph_aegnn/test_codes/data/ncars/training/sequence_1961',
 '/home/ibrahim/event_graph_aegnn/test_codes/data/ncars/training/sequence_5556',
 '/home/ibrahim/event_graph_aegnn/test_codes/data/ncars/training/sequence_9253',
 '/home/ibrahim/event_graph_aegnn/test_codes/data/ncars/training/sequence_10095',
 '/home/ibrahim/event_graph_aegnn/test_codes/data/ncars/training/sequence_9350',
 '/home/ibrahim/event_graph_aegnn/test_codes/data/ncars/training/sequence_6639',
 '/home/ibrahim/event_graph_aegnn/test_codes/data/ncars/training/sequence_10368',
 '/home/ibrahim/event_graph_aegnn/test_codes/data/ncars/training/sequence_7203',
 '/home/ibrahim/event_graph_aegnn/test_codes/data/ncars/training/sequence_762',
 '/home/ibrahim/event_graph_aegnn/test_codes/data/ncars/training/sequence_6426',
 '/home/ibrahim/event_graph_aegnn/test_codes/data/ncars/training/sequence_1168',
 '/home/ibrahim/event_graph_aegnn/test_codes/data/ncars/training/sequence_1384',
 '/home/ibrahim/event_graph

In [ ]:
ncars.classes

['car', 'background']

In [ ]:
ncars.pre_transform

<bound method NCars.pre_transform of NCars[Train: None
Validation: None]>

In [ ]:
#Loading raw event data from file name
event_sample = ncars.load(event_sample_file[0])
event_sample

Data(x=[5715, 4], pos=[5715, 3])

On a ici environ 10000 events dans le raw event cloud.
Chaque event est un noeud, ayant une feature (x) avec la polarité et une position spatiale (pos) avec les coordonnées (x,y,t) 

In [ ]:
event_sample.label = ncars.read_label(event_sample_file[0])
event_sample

Data(x=[5715, 4], pos=[5715, 3], label='background')

In [ ]:
class_dict = {class_id: i for i, class_id in enumerate(ncars.classes)}
event_sample.y = torch.tensor([class_dict[event_sample.label]])
event_sample


Data(x=[5715, 4], pos=[5715, 3], label='background', y=[1])

In [ ]:
event_sample_data_pt = ncars.pre_transform(event_sample)
event_sample_data_pt

Data(x=[5715, 4], pos=[5715, 3], label='background', y=[1], edge_index=[2, 151044])

### Pre-transform

In [ ]:
from aegnn.datasets.utils.normalization import normalize_time
from torch_geometric.nn.pool import radius_graph

In [ ]:
event_sample.pos

tensor([[1.7000e+01, 2.5000e+01, 0.0000e+00],
        [7.6000e+01, 4.2000e+01, 5.0000e-11],
        [1.7000e+01, 5.0000e+00, 7.0000e-11],
        ...,
        [2.4000e+01, 2.9000e+01, 4.9942e-07],
        [2.6000e+01, 1.4000e+01, 4.9963e-07],
        [2.5000e+01, 3.0000e+01, 4.9964e-07]])

In [ ]:
event_sample.pos.shape

torch.Size([5715, 3])

In [ ]:
torch.max(event_sample.pos[:,2])

tensor(4.9964e-07)

In [ ]:
event_sample.pos[:, 2] = normalize_time(event_sample.pos[:, 2])
event_sample.pos

tensor([[1.7000e+01, 2.5000e+01, 0.0000e+00],
        [7.6000e+01, 4.2000e+01, 2.5000e-16],
        [1.7000e+01, 5.0000e+00, 3.5000e-16],
        ...,
        [2.4000e+01, 2.9000e+01, 2.4971e-12],
        [2.6000e+01, 1.4000e+01, 2.4981e-12],
        [2.5000e+01, 3.0000e+01, 2.4982e-12]])

La coordonnée temporelle a été normalisée 

In [ ]:
params = ncars.hparams.preprocessing
event_subsampling = ncars.sub_sampling(event_sample, n_samples=params["n_samples"], sub_sample=params["sampling"])
event_subsampling

Data(x=[5715, 4], pos=[5715, 3], label='background', y=[1])

Le sub samplng de la class NCars permet de ramener le nombre d'events à exactement 10000. L'article indiquait un uniforme subsampling de fecteur K, ce qui n'est pas le cas ici. 

In [ ]:
event_subsampling.edge_index = radius_graph(event_subsampling.pos, r=params["r"], max_num_neighbors=params["d_max"])
event_subsampling

Data(x=[5715, 4], pos=[5715, 3], label='background', y=[1], edge_index=[2, 176728])

Une fois le pre-transform réalisé, il est stocké dans un fichier. C'est donc ici que s'arrête le fichier de preprocessing.py

## NCaltech

In [ ]:
from aegnn.datasets.ncaltech101 import NCaltech101

In [ ]:
ncaltech = NCaltech101()

In [ ]:
ncaltech

NCaltech101[Train: None
Validation: None]

In [ ]:
event_sample_file = ncaltech.raw_files("training")[0]
event_sample_file

IndexError: list index out of range

In [ ]:
ncaltech.classes

FileNotFoundError: [Errno 2] No such file or directory: '/home/ibrahim/event_graph_aegnn/test_codes/data/ncaltech101/training'

### Load

In [ ]:
import numpy as np

In [ ]:
f = open(event_sample_file, 'rb')
raw_data = np.fromfile(f, dtype=np.uint8)
print(raw_data)
f.close()
raw_data = np.uint32(raw_data)

TypeError: unhashable type: 'list'

In [ ]:
all_y = raw_data[1::5]
all_x = raw_data[0::5]
all_p = (raw_data[2::5] & 128) >> 7 

NameError: name 'raw_data' is not defined

In [ ]:
raw_data[1::5]

NameError: name 'raw_data' is not defined

In [ ]:
raw_data.shape

NameError: name 'raw_data' is not defined

In [ ]:
((raw_data[2::5] & 127) << 16).shape

NameError: name 'raw_data' is not defined

In [ ]:
((raw_data[3::5] << 8)).shape

NameError: name 'raw_data' is not defined

In [ ]:
(raw_data[4::5]).shape

NameError: name 'raw_data' is not defined

In [ ]:
all_ts = ((raw_data[2::5] & 127) << 16) | (raw_data[3::5] << 8) | (raw_data[4::5])

NameError: name 'raw_data' is not defined

In [ ]:
class_dict = {class_id: i for i, class_id in enumerate(ncaltech.classes)}

FileNotFoundError: [Errno 2] No such file or directory: '/home/ibrahim/event_graph_aegnn/test_codes/data/ncaltech101/training'

In [ ]:
class_dict

{'car': 0, 'background': 1}

In [ ]:
event_sample = ncaltech.load(event_sample_file)
event_sample

TypeError: expected str, bytes or os.PathLike object, not list

In [ ]:
label = ncaltech.read_label(event_sample_file)
label

AttributeError: 'list' object has no attribute 'split'

In [ ]:
event_sample.label = label if isinstance(label, list) else [label]
event_sample.y = torch.tensor([class_dict[label] for label in event_sample.label])

NameError: name 'label' is not defined

In [ ]:
bbox = ncaltech.read_annotations(event_sample_file)
bbox

TypeError: expected str, bytes or os.PathLike object, not list

In [ ]:
event_sample.bbox = torch.tensor(bbox, device="cuda").long()

NameError: name 'bbox' is not defined

In [ ]:
event_sample

Data(x=[5715, 4], pos=[5715, 3], label='background', y=[1])

data_transform

In [ ]:
window_us = 50 * 1000
t = event_sample.pos[event_sample.num_nodes // 2, 2]
t

tensor(1.2884e-12)

In [ ]:
data = event_sample
event_sample

Data(x=[5715, 4], pos=[5715, 3], label='background', y=[1])

In [ ]:
num_nodes = data.num_nodes
num_nodes

5715

In [ ]:
index1 = torch.clamp(torch.searchsorted(data.pos[:, 2].contiguous(), t) - 1, 0, data.num_nodes - 1)
index0 = torch.clamp(torch.searchsorted(data.pos[:, 2].contiguous(), t-window_us) - 1, 0, data.num_nodes - 1)

In [ ]:
for key, item in data:
            if torch.is_tensor(item) and item.size(0) == num_nodes and item.size(0) != 1:
                data[key] = item[index0:index1, :]

In [ ]:
torch.is_tensor(data.pos)

True

In [ ]:
data.pos.size(0)

2856

In [ ]:
data

Data(x=[2856, 4], pos=[2856, 3], label='background', y=[1])

In [ ]:
params = ncaltech.hparams.preprocessing

In [ ]:
data = ncaltech.sub_sampling(data, n_samples=params["n_samples"], sub_sample=params["sampling"])

In [ ]:
data

Data(x=[2856, 4], pos=[2856, 3], label='background', y=[1])

Le dataset Ncaltech101, est à la fois un dataset de classification et de détection d'objets. Niveau annotation, il n'y a qu'une seule box, indépendente du temps, ce qui rend la tâche plus simple. Pour un sample, l'article choisie une fenêtre temporelle de 50 microsecondes. On récuper le temps de l'event T qui correspond à la moitié du sample. On prend comme index les indices de la fenêtre temporelle juste avant T. On réalise ensuite un subsampling sur 

In [ ]:
processed_files = ncaltech.processed_files("training")
processed_files

[]

In [ ]:
ncaltech._load_dataset("training")

In [ ]:
ncaltech.setup()

In [ ]:
g = ncaltech.train_dataset[0]

In [ ]:
g.bbox

tensor([[  5,   4, 173, 146,   0]])

# Training

## Dataset et Dataloader

In [ ]:
from torch_geometric.transforms import Cartesian
from torch_geometric.data import Data
import torch_geometric

In [ ]:
processed_files = ncars.processed_files("training")
processed_files

['/home/ibrahim/event_graph_aegnn/test_codes/data/ncars/processed/training/sequence_1961',
 '/home/ibrahim/event_graph_aegnn/test_codes/data/ncars/processed/training/sequence_5556',
 '/home/ibrahim/event_graph_aegnn/test_codes/data/ncars/processed/training/sequence_9253',
 '/home/ibrahim/event_graph_aegnn/test_codes/data/ncars/processed/training/sequence_10095',
 '/home/ibrahim/event_graph_aegnn/test_codes/data/ncars/processed/training/sequence_9350',
 '/home/ibrahim/event_graph_aegnn/test_codes/data/ncars/processed/training/sequence_6639',
 '/home/ibrahim/event_graph_aegnn/test_codes/data/ncars/processed/training/sequence_10368',
 '/home/ibrahim/event_graph_aegnn/test_codes/data/ncars/processed/training/sequence_7203',
 '/home/ibrahim/event_graph_aegnn/test_codes/data/ncars/processed/training/sequence_762',
 '/home/ibrahim/event_graph_aegnn/test_codes/data/ncars/processed/training/sequence_6426',
 '/home/ibrahim/event_graph_aegnn/test_codes/data/ncars/processed/training/sequence_1168'

In [ ]:
event_processed_file = processed_files[0]
event_processed_file

'/home/ibrahim/event_graph_aegnn/test_codes/data/ncars/processed/training/sequence_1961'

In [ ]:
event_processed_data = torch.load(event_processed_file, weights_only=False)
event_processed_data

Data(x=[5715, 4], pos=[5715, 3], file_id='sequence_1961', label=[1], y=[1], edge_index=[2, 151044])

Le dataset Ncars étant un dataset de classification, il n'y a pas de bounding box. Il n'y a pas de transform non plus.  

In [ ]:
max_value = ncars.hparams.get("preprocessing", {}).get("r", None)
edge_attr_func = Cartesian(norm=True, max_value=max_value, cat=False)
edge_attr_func(event_processed_data)
event_processed_data

Data(x=[5715, 4], pos=[5715, 3], file_id='sequence_1961', label=[1], y=[1], edge_index=[2, 151044])

Le dataset est donc créé à partir de ces données.

## Create and run model


### Construction graphe

In [ ]:
dataset = data_module.train_dataset

In [ ]:
index = 0
num_events = 25000
radius = 3.0
max_num_neighbors = 32


In [ ]:
sample = dataset[index % len(dataset)]
sample

Data(x=[5715, 4], pos=[5715, 3], file_id='sequence_1961', label=[1], y=[1], edge_index=[2, 151044], edge_attr=[151044, 3])

In [ ]:
sample.pos = sample.pos[:, :2] #Pourquoi retirer la composante temporelle de la position?
sample

Data(x=[5715, 4], pos=[5715, 2], file_id='sequence_1961', label=[1], y=[1], edge_index=[2, 151044], edge_attr=[151044, 3])

In [ ]:
data = Data(x=sample.x[:num_events], pos=sample.pos[:num_events])
data.batch = torch.zeros(data.num_nodes, device=data.x.device)
data.edge_index = torch_geometric.nn.radius_graph(data.pos, r=radius, max_num_neighbors=max_num_neighbors).long()
data.edge_attr = edge_attr_func(data).edge_attr
data

Data(x=[5715, 4], pos=[5715, 2], batch=[5715], edge_index=[2, 176728], edge_attr=[176728, 2])

In [ ]:
index_new = min(num_events, sample.num_nodes - 1)
x_new = sample.x[index_new, :].view(1, -1)
pos_new = sample.pos[index_new, :2].view(1, -1)
event_new = Data(x=x_new, pos=pos_new, batch=torch.zeros(1, dtype=torch.long))
event_new

Data(x=[1, 4], pos=[1, 2], batch=[1])

`data` est le graphe initial. `event_new` se rajoute au graphe initial de manière asynchrone.

### Create model

In [ ]:
input_shape = torch.tensor([*data_module.dims, data.pos.shape[-1]], device='cpu')
pooling_size = (10,10)

In [ ]:
model = aegnn.models.networks.GraphRes(data_module.name, input_shape, data_module.num_classes, pooling_size=pooling_size)

In [ ]:
model._modules.items()

dict_items([('conv1', SplineConv(4, 8, dim=2)), ('norm1', BatchNorm(8, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)), ('conv2', SplineConv(8, 16, dim=2)), ('norm2', BatchNorm(16, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)), ('conv3', SplineConv(16, 16, dim=2)), ('norm3', BatchNorm(16, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)), ('conv4', SplineConv(16, 16, dim=2)), ('norm4', BatchNorm(16, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)), ('conv5', SplineConv(16, 32, dim=2)), ('norm5', BatchNorm(32, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)), ('pool5', MaxPooling(voxel_size=[10, 10])), ('conv6', SplineConv(32, 32, dim=2)), ('norm6', BatchNorm(32, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)), ('conv7', SplineConv(32, 32, dim=2)), ('norm7', BatchNorm(32, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)), ('pool7', MaxPoolingX(voxel_size=tensor([30, 25]), s

In [ ]:
model

GraphRes(
  (conv1): SplineConv(4, 8, dim=2)
  (norm1): BatchNorm(8, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (conv2): SplineConv(8, 16, dim=2)
  (norm2): BatchNorm(16, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (conv3): SplineConv(16, 16, dim=2)
  (norm3): BatchNorm(16, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (conv4): SplineConv(16, 16, dim=2)
  (norm4): BatchNorm(16, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (conv5): SplineConv(16, 32, dim=2)
  (norm5): BatchNorm(32, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (pool5): MaxPooling(voxel_size=[10, 10])
  (conv6): SplineConv(32, 32, dim=2)
  (norm6): BatchNorm(32, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (conv7): SplineConv(32, 32, dim=2)
  (norm7): BatchNorm(32, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (pool7): MaxPoolingX(voxel_size=tensor([30, 25]), size=16)
  (fc): 

In [ ]:
model.to("cpu")

GraphRes(
  (conv1): SplineConv(4, 8, dim=2)
  (norm1): BatchNorm(8, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (conv2): SplineConv(8, 16, dim=2)
  (norm2): BatchNorm(16, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (conv3): SplineConv(16, 16, dim=2)
  (norm3): BatchNorm(16, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (conv4): SplineConv(16, 16, dim=2)
  (norm4): BatchNorm(16, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (conv5): SplineConv(16, 32, dim=2)
  (norm5): BatchNorm(32, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (pool5): MaxPooling(voxel_size=[10, 10])
  (conv6): SplineConv(32, 32, dim=2)
  (norm6): BatchNorm(32, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (conv7): SplineConv(32, 32, dim=2)
  (norm7): BatchNorm(32, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (pool7): MaxPoolingX(voxel_size=tensor([30, 25]), size=16)
  (fc): 

### Make asynchronous

#### Modèle entier

In [ ]:
log_flops=True
log_runtime=True

In [ ]:
async_model = aegnn.asyncronous.make_model_asynchronous(model, radius, list(data_module.dims), edge_attr_func, log_flops=log_flops, log_runtime=log_runtime)

In [ ]:
async_model

GraphRes(
  (conv1): SplineConv(4, 8, dim=2)
  (norm1): BatchNorm(8, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (conv2): SplineConv(8, 16, dim=2)
  (norm2): BatchNorm(16, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (conv3): SplineConv(16, 16, dim=2)
  (norm3): BatchNorm(16, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (conv4): SplineConv(16, 16, dim=2)
  (norm4): BatchNorm(16, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (conv5): SplineConv(16, 32, dim=2)
  (norm5): BatchNorm(32, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (pool5): MaxPooling(voxel_size=[10, 10])
  (conv6): SplineConv(32, 32, dim=2)
  (norm6): BatchNorm(32, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (conv7): SplineConv(32, 32, dim=2)
  (norm7): BatchNorm(32, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (pool7): MaxPoolingX(voxel_size=tensor([30, 25]), size=16)
  (fc): 

In [ ]:
async_model.conv1.asy_pos

#### Asynchronous Conv

In [ ]:
from torch_geometric.nn.conv import SplineConv

In [ ]:
model = aegnn.models.networks.GraphRes(data_module.name, input_shape, data_module.num_classes, pooling_size=pooling_size)

In [ ]:
spline_conv = SplineConv(1, 8, dim=2, kernel_size=2, bias=False, root_weight=False)

In [ ]:
async_conv_layer = model.conv1
async_conv_layer

SplineConv(4, 8, dim=2)

In [ ]:
# rajoute une mémoire du graphe à la couche de convolution
async_conv_layer.asy_graph = None
async_conv_layer.asy_flops_log = []
async_conv_layer.asy_runtime_log = []
async_conv_layer.asy_radius = radius

In [ ]:
async_conv_layer.asy_pos = None

In [ ]:
from aegnn.asyncronous.base.base import make_asynchronous

In [ ]:
async_conv_layer.asy_is_initial = True #Pour la première couche de convoulution, sinon False
async_conv_layer.asy_edge_attributes = edge_attr_func #Fonction pour calculer les attributs
async_conv_layer.sync_forward = async_conv_layer.forward #Forward synchrone

Appel de la fonction make_synchronous() avec les arguments __graph_initialisation et __graph_processing

La fonction make_asynchronous permet de changer le forward du module en passant de synchrone à asynchrone. La fonction async_forward() appelle async_context() qui détermine quel func de graphe est appellé, suivant si on est sur le graphe d'initialisation ou sur un nouvel event. Cette fonction permet de calculer le temps d'execution pour le traitement d'un event. 

Une fois la fonction async_forward() définie comme module.forward(), une fonction de callback est défini. Celle-ci permet d'affecter à un attribut d'une couche suivante dans le modèle, la valeur de cette attribut pour la couche actuelle.

#### Forward

##### __graph_initialisation

In [ ]:
#Lors de l'appel du forward, l'attribut pos du graphe passé en argument est assigné à l'attribut asy_pos de chaque couche (hors BatchNormalisation)
async_model.asy_pass_attribute('asy_pos', data.pos)

In [ ]:
async_conv_layer = async_model.conv1
async_conv_layer.asy_edge_attributes

Cartesian(norm=True, max_value=3.0)

In [ ]:
pos = async_conv_layer.asy_pos
print(pos)

tensor([[25., 13.],
        [83., 47.],
        [19., 28.],
        ...,
        [19., 43.],
        [24., 13.],
        [20., 20.]])


On suppose ici que les edge_attributs et les egdes index sont donnés à la couche (ce qui est le cas dans le code) 

In [ ]:
#Le graphe initiale est inféré avec la fonction forward synchrone
y = async_conv_layer.sync_forward(data.x, edge_index=data.edge_index, edge_attr=data.edge_attr)
y.shape

/tmp/ipykernel_311714/1008329618.py:2: UserWarning: We do not recommend using the non-optimized CPU version of `SplineConv`. If possible, please move your data to GPU.
  y = async_conv_layer.sync_forward(data.x, edge_index=data.edge_index, edge_attr=data.edge_attr)


torch.Size([5715, 8])

In [ ]:
async_conv_layer.asy_graph = Data(x=data.x, pos=pos, edge_index=data.edge_index, edge_attr=data.edge_attr,y=y)
async_conv_layer.asy_graph

Data(x=[5715, 4], edge_index=[2, 176728], edge_attr=[176728, 2], y=[5715, 8], pos=[5715, 2])

Une instruction permet de calculer le nombre de flops nécessaire à l'inférence du graphe d'initialisation

In [ ]:
#Fonction qui permet d'assigner pos à l'attribut asy_pos des couches suivantes dans le réseau
#Semble faire doublon avec 
async_conv_layer.asy_pass_attribute('asy_pos', pos)

#### __graph_processing

Une fois le graph d'initialisation inféré, c'est la fonction __graph_processing qui est appelé pour le forward. La fonction prend en argument le module (la couche elle même), x (les features de la nouvelle node), edge_index et edge_attributes 

##### Conv1 couche initial

In [ ]:
async_model.asy_pass_attribute('asy_pos', event_new.pos)

In [ ]:
async_conv_layer.asy_pos

tensor([[20., 20.]])

In [ ]:
pos = async_conv_layer.asy_pos

In [ ]:
print(f"Input graph with x = {event_new.x.shape} and pos = {pos.shape}")
print(f"Internal graph = {async_conv_layer.asy_graph}")

Input graph with x = torch.Size([1, 4]) and pos = torch.Size([1, 2])
Internal graph = Data(x=[5715, 4], edge_index=[2, 176728], edge_attr=[176728, 2], y=[5715, 8], pos=[5715, 2])


On étudie la couche initiale ici (conv1)

In [ ]:
num_prev_nodes = async_conv_layer.asy_graph.num_nodes
num_prev_nodes

5715

In [ ]:
x_new, pos_new = event_new.x, pos

In [ ]:
#Détermine les index des nouvelles nodes après concaténation  
idx_new = torch.arange(num_prev_nodes, num_prev_nodes + event_new.x.size()[0], device=event_new.x.device)
print(idx_new)

tensor([5715])


In [ ]:

idx_diff = torch.tensor([], device=event_new.x.device, dtype=torch.long)
idx_diff

tensor([], dtype=torch.int64)

In [ ]:
x_all = torch.cat([async_conv_layer.asy_graph.x, x_new], dim=0)
pos_all = torch.cat([async_conv_layer.asy_graph.pos, pos_new], dim=0)

In [ ]:
print(f"Subgraph contains {idx_new.numel()} new and {idx_diff.numel()} diff nodes")

Subgraph contains 1 new and 0 diff nodes


In [ ]:
#Détermine les nodes existants qui sont connectés avec la nouvelle node
connected_node_mask = torch.cdist(pos_all, pos_new) <= async_conv_layer.asy_radius
connected_node_mask.sum()

tensor(314)

In [ ]:
torch.cdist(pos_all, pos_new) 

In [ ]:
idx_new_neigh = torch.unique(torch.nonzero(connected_node_mask)[:, 0])
idx_new_neigh #Index des voisins de la nouvelle node

tensor([   15,   108,   140,   142,   143,   145,   190,   247,   278,   301,
          334,   362,   402,   443,   475,   487,   491,   497,   513,   524,
          550,   605,   619,   629,   710,   739,   765,   801,   877,   892,
          902,   906,   929,   980,   993,   995,  1017,  1020,  1072,  1104,
         1142,  1151,  1155,  1181,  1205,  1206,  1227,  1241,  1244,  1280,
         1299,  1332,  1339,  1350,  1370,  1398,  1441,  1468,  1474,  1490,
         1510,  1520,  1527,  1532,  1559,  1652,  1666,  1703,  1712,  1745,
         1757,  1773,  1778,  1794,  1853,  1889,  1941,  1980,  2036,  2164,
         2190,  2203,  2271,  2280,  2400,  2402,  2410,  2444,  2452,  2462,
         2507,  2519,  2525,  2541,  2601,  2685,  2856,  2922,  2942,  2953,
         2956,  2970,  2987,  2988,  3018,  3063,  3185,  3222,  3321,  3404,
         3442,  3451,  3469,  3484,  3489,  3504,  3519,  3561,  3655,  3678,
         3704,  3731,  3749,  3793,  3815,  3873,  3882,  3904, 

In [ ]:
idx_update = torch.cat([idx_new_neigh, idx_diff]) 
#Les index des noeuds impactés depuis le dernier forward (noeuds dont les features ont changé (aucune a priori) et ceux recevant un nouveau voisin)

In [ ]:
idx_update.shape

torch.Size([314])

In [ ]:
from torch_geometric.utils import k_hop_subgraph
# récupère les voisins des voisins de la nouvelle node
_, edges_connected, _, connected_edges_mask = k_hop_subgraph(idx_update, num_hops=1,
                                                                 edge_index=async_conv_layer.asy_graph.edge_index,
                                                                 num_nodes=pos_all.shape[0])

In [ ]:
edges_new = torch.nonzero(connected_node_mask).T
edges_new[1, :] = idx_new[edges_new[1,:]] #changement d'index pour ramener à l'index du graphe global

In [ ]:
edges_new_inv = torch.stack([edges_new[1, :], edges_new[0, :]], dim=0)
edges_new_inv.shape

torch.Size([2, 314])

In [ ]:
edges_new = torch.cat([edges_new, edges_new_inv], dim=1)
edges_new.shape
#Permet d'obtenir un graphe non orienté

torch.Size([2, 628])

In [ ]:
edges_new = torch.unique(edges_new, dim=1)
edges_new.shape

torch.Size([2, 627])

In [ ]:
from torch_geometric.utils import remove_self_loops
edges_new, _ = remove_self_loops(edges_new)

In [ ]:
edge_index = torch.cat([edges_connected, edges_new], dim=1) # On obtient les edges index d'un subgraph de k = 2 à partir de la nouvelle node

In [ ]:
async_conv_layer.asy_edge_attributes

Cartesian(norm=True, max_value=3.0)

In [ ]:
#Calul des edges_attributes
graph_new = Data(x=x_all, pos=pos_all, edge_index=edges_new)
edge_attr_new = async_conv_layer.asy_edge_attributes(graph_new).edge_attr
edge_attr_connected = async_conv_layer.asy_graph.edge_attr[connected_edges_mask, :]
edge_attr = torch.cat([edge_attr_connected, edge_attr_new])
edge_attr

tensor([[0.5000, 0.5000],
        [0.5000, 0.5000],
        [0.5000, 0.5000],
        ...,
        [0.3333, 0.8333],
        [0.6667, 0.8333],
        [0.5000, 0.5000]])

In [ ]:
x_new

tensor([[0.]])

In [ ]:
out_channels = async_conv_layer.asy_graph.y.size()[-1]
y = torch.cat([async_conv_layer.asy_graph.y.clone(), torch.zeros(x_new.size()[0], out_channels, device=event_new.x.device)])

In [ ]:
x_j = x_all[edge_index[0, :], :] #Features des voisins et de distance k-hops = 2
x_j

tensor([[0.],
        [0.],
        [0.],
        ...,
        [0.],
        [0.],
        [0.]])

In [ ]:
#Calcule du message 
phi = async_conv_layer.message(x_j, edge_attr=edge_attr)

In [ ]:
y_update = async_conv_layer.aggregate(phi, index=edge_index[1,:], ptr=None, dim_size=x_all.size()[0])
y_update

tensor([[0.0000, 0.0000, 0.0000,  ..., 0.0000, 0.0000, 0.0000],
        [0.0000, 0.0000, 0.0000,  ..., 0.0000, 0.0000, 0.0000],
        [0.0000, 0.0000, 0.0000,  ..., 0.0000, 0.0000, 0.0000],
        ...,
        [0.0000, 0.0000, 0.0000,  ..., 0.0000, 0.0000, 0.0000],
        [0.0000, 0.0000, 0.0000,  ..., 0.0000, 0.0000, 0.0000],
        [0.0003, 0.0042, 0.0040,  ..., 0.0006, 0.0003, 0.0037]],
       grad_fn=<DivBackward0>)

In [ ]:
y[idx_update] = y_update[idx_update]

In [ ]:
print(f"Updated {idx_update.numel()} nodes in asy. graph of module {async_conv_layer}")

Updated 314 nodes in asy. graph of module SplineConv(1, 8, dim=2)


Il y a possibilté de calculer le nomnbre de flops si besoin. Si la couche est une couche initiale (comme conv1), on assigne à l'attribut asy_pos des autres couches la valeur pos_all 

##### Conv2

In [ ]:
data

Data(x=[10000, 1], pos=[10000, 2], batch=[10000], edge_index=[2, 315712], edge_attr=[315712, 2])

In [ ]:
model = aegnn.models.networks.GraphRes(data_module.name, input_shape, data_module.num_classes, pooling_size=pooling_size)
async_model = aegnn.asyncronous.make_model_asynchronous(model, radius, list(data_module.dims), edge_attr_func, log_flops=log_flops, log_runtime=log_runtime)

In [ ]:
_ = async_model(data)

In [ ]:
event_new

Data(x=[1, 1], pos=[1, 2], batch=[1])

In [ ]:
from torch.nn.functional import elu

In [ ]:
async_model.asy_pass_attribute('asy_pos', event_new.pos)
x_f = elu(async_model.conv1(event_new.x))
x_f = async_model.norm1(x_f)

In [ ]:
async_conv2_layer = async_model.conv2

On se retouve maintenant dans la partie qui concerne les couches de convolution qui ne sont pas la première

In [ ]:
pos = async_conv2_layer.asy_pos

In [ ]:
print(f"Input graph with x = {x_f.shape} and pos = {pos.shape}")
print(f"Internal graph = {async_conv2_layer.asy_graph}")

Input graph with x = torch.Size([10001, 8]) and pos = torch.Size([10001, 2])
Internal graph = Data(x=[10000, 8], edge_index=[2, 315712], edge_attr=[315712, 2], y=[10000, 16], pos=[10000, 2])


Pour la couche de convolution 2, l'input est la sortie de la conv1, avec donc donc 10001 nodes de 8 channels chacun.

In [ ]:
from aegnn.asyncronous.base.utils import graph_new_nodes, graph_changed_nodes

In [ ]:
#On récupère la nouvelle node et son index dans le graphe
x_new, idx_new = graph_new_nodes(async_conv2_layer, x=x_f)

In [ ]:
pos_new = pos[idx_new, :] #La position de la nouvelle node

In [ ]:
_, idx_diff = graph_changed_nodes(async_conv2_layer, x=x_f) #On récupère les index des nodes qui ont changé (qui ont été impacté par l'ajout de la nouvelle node dans la conv1)

In [ ]:
idx_diff

tensor([ 142,  145,  475,  487,  497,  550,  980,  993, 1072, 1104, 1151, 1559,
        1652, 1666, 1778, 1794, 1889, 1980, 2164, 2190, 2400, 2410, 2856, 2922,
        2956, 2970, 3018, 3063, 3222, 3442, 3484, 3655, 3678, 3704, 3731, 3793,
        3882, 4071, 4077, 4179, 4321, 4374, 5505, 5515, 5862, 6036, 6254, 6289,
        6685, 6790, 6802, 7145, 7326, 7487, 7572, 7604, 7971, 8200, 8465, 8526,
        8873, 9032, 9048, 9087, 9323, 9326, 9552, 9600, 9763])

In [ ]:
#On récupère les indexs des voisins des noeuds qui ont changés
idx_diff, _, _, _ = k_hop_subgraph(idx_diff, num_hops=1, edge_index=async_conv2_layer.asy_graph.edge_index,
                                               num_nodes=async_conv2_layer.asy_graph.num_nodes + len(idx_new))

La suite est le même traitement pour trouver les edge_index et les edge_attr

### Etude de la couche Batch Normalization

Les couches de BN peuvent être rendu asynchrone. Contrairement aux autres types de couches, celles-ci n'ont pas de callback et ne sont pas concernés par le callback des autres.

In [ ]:
from torch.nn.functional import elu

In [ ]:
data

Data(x=[10000, 1], pos=[10000, 2], batch=[10000], edge_index=[2, 315712], edge_attr=[315712, 2])

In [ ]:
model = aegnn.models.networks.GraphRes(data_module.name, input_shape, data_module.num_classes, pooling_size=pooling_size)

In [ ]:
async_bn_layer = model.norm1

In [ ]:
async_model = aegnn.asyncronous.make_model_asynchronous(model, radius, list(data_module.dims), edge_attr_func, log_flops=log_flops, log_runtime=log_runtime)

In [ ]:
async_model.asy_pass_attribute('asy_pos', data.pos)

In [ ]:
x_f = elu(async_model.conv1(data.x, data.edge_index, data.edge_attr))

In [ ]:
async_bn_layer.asy_graph = None
async_bn_layer.asy_flops_log = [] if log_flops else None
async_bn_layer.asy_runtime_log = [] if log_runtime else None


#### __graph_initiatization

In [ ]:
#On calcule la moyenne et la variance du graphe initiale
mean = torch.mean(x_f, dim = 0)
var = torch.var(x_f, dim = 0) + async_bn_layer.module.eps

In [ ]:
async_bn_layer.asy_graph = torch_geometric.data.Data(x=mean,variance=var)
#La moyenne et la variance sont stocké dans un graphe

#### _graph_processing

In [ ]:
model = aegnn.models.networks.GraphRes(data_module.name, input_shape, data_module.num_classes, pooling_size=pooling_size)
async_model = aegnn.asyncronous.make_model_asynchronous(model, radius, list(data_module.dims), edge_attr_func, log_flops=log_flops, log_runtime=log_runtime)
_ = async_model(data)

In [ ]:
async_bn_layer = async_model.norm1

In [ ]:
async_model.asy_pass_attribute('asy_pos', event_new.pos)
x_f = elu(async_model.conv1(event_new.x))

In [ ]:
x_f.shape

torch.Size([10001, 8])

In [ ]:
y = (x_f - async_bn_layer.asy_graph.x) / async_bn_layer.asy_graph.variance * async_bn_layer.module.weight + async_bn_layer.module.bias

On normalise puis on scale les features des noeuds du graphe en entrée

### Etude de la couche linéaire

In [ ]:
model = aegnn.models.networks.GraphRes(data_module.name, input_shape, data_module.num_classes, pooling_size=pooling_size)
async_model = aegnn.asyncronous.make_model_asynchronous(model, radius, list(data_module.dims), edge_attr_func, log_flops=log_flops, log_runtime=log_runtime)
_ = async_model(data)

In [ ]:
async_model.feature_map.shape

torch.Size([1, 512])

In [ ]:
async_model(event_new)

tensor([[-32.7765,  -6.9884]], grad_fn=<CopySlices>)

In [ ]:
async_linear_layer = async_model.fc
feature_map = async_model.feature_map

In [ ]:
async_linear_layer.asy_graph

Data(x=[1, 512], y=[1, 2])

In [ ]:
feature_map.shape

torch.Size([1, 512])

In [ ]:
diff_idx = torch.nonzero(feature_map - async_linear_layer.asy_graph.x, as_tuple=True)

In [ ]:
len(diff_idx) * diff_idx[0].shape[0]

0

In [ ]:
diff_idx2 = torch.nonzero(feature_map - async_linear_layer.asy_graph.x).t().detach().cpu().numpy()
diff_idx2.shape[0] * diff_idx2.shape[1] 

0

In [ ]:
feature_map[diff_idx]

tensor([], grad_fn=<IndexBackward0>)

In [ ]:
async_linear_layer.asy_graph.x[diff_idx]

tensor([], grad_fn=<IndexBackward0>)

In [ ]:
x_diff = feature_map[diff_idx] - async_linear_layer.asy_graph.x[diff_idx]

In [ ]:
#On calcule un résidu sur les features qui ont changés depuis l'ajout d'un nouvel event
y_residual = torch.mul(async_linear_layer.weight[:, diff_idx[1]], x_diff).t()

In [ ]:
async_linear_layer.weight.shape

torch.Size([2, 512])

In [ ]:
#update
async_linear_layer.asy_graph.x[diff_idx] = feature_map[diff_idx]


In [ ]:
async_linear_layer.asy_graph.y[diff_idx[0],:] += y_residual

In [ ]:
async_linear_layer.asy_graph.y

tensor([[-32.7765,  -6.9884]], grad_fn=<CopySlices>)

In [ ]:
flops = int(diff_idx[0].shape[0] * diff_idx[1].shape[0])
flops

0

In [ ]:
diff_idx[0].shape[0]

0

In [ ]:
diff_idx[1].shape

torch.Size([0])

### Etude des couches de pooling et de l'asynchronicité

Dans le modèle AEGNN, il y a 2 types de couches pour le max pooling.
- max_pooling: des voxels spaciaux sont créés à partir de la position des noeuds et du voxel size puis un max pooling est réalisé sur chaque cluster. A noter que la composante temporelle ne fait pas partie de la clusterisation. 
- max_pooling_x : 

##### max_pooling layer

In [ ]:
from torch_geometric.nn.pool import max_pool, voxel_grid

In [ ]:
model = aegnn.models.networks.GraphRes(data_module.name, input_shape, data_module.num_classes, pooling_size=pooling_size)
async_model = aegnn.asyncronous.make_model_asynchronous(model, radius, list(data_module.dims), edge_attr_func, log_flops=log_flops, log_runtime=log_runtime)
_ = async_model(data)

In [ ]:
data_before_pool= async_model.fm_to_pool

In [ ]:
data_before_pool

Data(x=[10000, 32], edge_index=[2, 315712], edge_attr=[315712, 2], pos=[10000, 2], batch=[10000])

In [ ]:
cluster = voxel_grid(data_before_pool.pos[:, :2], batch = data_before_pool.batch, size = pooling_size)
cluster.shape

torch.Size([10000])

In [ ]:
torch.unique(cluster, return_counts=True)

(tensor([ 0,  1,  2,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17, 18,
         19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29]),
 tensor([  1, 301,   4, 103, 638, 599, 710, 785, 334, 861, 295, 326, 322, 336,
         810, 275, 266, 287, 240, 434,  44,  40,  52,  92, 373, 643, 336, 266,
         227]))

In [ ]:
k = 1 #ID du voxel
mask = (cluster == k)

nodes_voxel_index = mask.nonzero(as_tuple=True)[0]
nodes_voxel_index.shape

torch.Size([301])

In [ ]:
voxel_graph = Data(x=data_before_pool.x[nodes_voxel_index], pos=data_before_pool.pos[nodes_voxel_index])
voxel_graph

Data(x=[301, 32], pos=[301, 2])

In [ ]:
max_pooled_features = torch.max(voxel_graph.x, dim=0)
max_pooled_features

torch.return_types.max(
values=tensor([16.2208, 26.2831, 21.3242, 18.8833, 11.5454, 17.7644, 28.4752, 10.0765,
        16.0865, 19.4614, 24.0271, 27.1238, 24.7543, 19.4208,  7.5912,  2.0817,
        14.7113,  9.2650, 18.4596, 29.5529, -1.7825, 14.0600, 24.1226,  9.8835,
        60.6269, 25.1525, 14.8999,  5.2461, 15.5529,  8.0962, 13.8667, 35.0324],
       grad_fn=<MaxBackward0>),
indices=tensor([260, 189,  41, 229,  40,  44,  18, 282, 119, 282,  41,  28,  30, 154,
         91,  28, 212,   4, 282,  92, 262,  30, 295, 116,  30, 143,  64,  40,
        128,  19,  58,  41]))

In [ ]:
transform = Cartesian(norm=True, cat=False)

In [ ]:
data_after_pool = max_pool(cluster, data=data_before_pool, transform=transform)
data_after_pool

DataBatch(x=[29, 32], edge_index=[2, 121], edge_attr=[121, 2], pos=[29, 2], batch=[29])

In [ ]:
unique_clusters = torch.unique(cluster)

idx = (unique_clusters == k).nonzero(as_tuple=True)[0].item()
idx

1

In [ ]:
pooled_nodes_features = data_after_pool.x[idx]

In [ ]:
torch.max(max_pooled_features.values - pooled_nodes_features)

tensor(0., grad_fn=<MaxBackward1>)

Pour chaque cluster, les features de la pooled node sont calculés en gardant le max sur chaque canal indépendement en prenant en compte les features des nodes contituant le cluster. 

In [ ]:
max_pool_pos = torch.mean(voxel_graph.pos, dim=0)
max_pool_pos

tensor([12.6047,  4.4452])

In [ ]:
torch.max(max_pool_pos - data_after_pool.pos[idx])

tensor(0.)

Pour chaque cluster, la position de la pooled node est calculée en faisant la moyenne des positions des nodes qui composent le cluster

Pour les edges_index, on considère un edge entre 2 clusters si une node de chaque cluster entre elles avaient un edge.
Les edges_attributes sont calculés ensuites comme pour les autres graphes

##### max_pooling_x

In [ ]:
from torch_geometric.nn.pool import max_pool_x

In [ ]:
data_before_pool_x = async_model.fm_to_pool_x

In [ ]:
voxel_size = input_shape[:2] // 4
voxel_size

tensor([30, 25])

In [ ]:
cluster = voxel_grid(data_before_pool_x.pos[:,:2], batch=data_before_pool_x.batch, size=voxel_size)
cluster

NameError: name 'voxel_grid' is not defined

In [ ]:
torch.unique(cluster, return_counts=True)

(tensor([0, 1, 2, 3, 4, 5]), tensor([9, 5, 6, 4, 3, 2]))

In [ ]:
#Seuls les features des noeuds sont renvoyés
#L'argument size permet d'obtenir le nombre de clusters renvoyés. Ici, Le nombre de cluster demandé (16) est plus grand que le nombre de cluster renvoyés par voxel grid
x_pooled, _ = max_pool_x(cluster, data_before_pool_x.x, data_before_pool_x.batch, size=16)
x_pooled.shape

torch.Size([16, 32])

Cette seconde couche de pooling est une version simplifié de la première qui ne renvoie que les features.

### Max pool asynchrone

In [ ]:
model = aegnn.models.networks.GraphRes(data_module.name, input_shape, data_module.num_classes, pooling_size=pooling_size)

In [ ]:
async_mp_layer = model.pool5

In [ ]:
async_mp_layer.asy_graph = None
async_mp_layer.asy_flops_log = [] if log_flops else None
async_mp_layer.asy_runtime_log = [] if log_runtime else None
async_mp_layer.asy_radius = radius

In [ ]:
async_mp_layer.asy_pos = None
async_mp_layer.asy_graph_coarse = None  # coarse output graph
async_mp_layer.asy_node_max_index = None  # index of max. node in input data
async_mp_layer.asy_voxel_pos_sum = None  # sum of positions per voxel
async_mp_layer.asy_voxel_node_count = None  # count of nodes per voxel

In [ ]:
async_mp_layer.grid_size = list(data_module.dims)# grid size in N dimensions

In [ ]:
async_mp_layer.grid_size

[120, 100]

La couche de max pooling garde en mémoire le coarse graph.

#### __graph_initialization

In [ ]:
from aegnn.asyncronous.max_pool import __dense_process, __get_clusters
from torch_scatter import scatter_max, scatter_sum

In [ ]:
model = aegnn.models.networks.GraphRes(data_module.name, input_shape, data_module.num_classes, pooling_size=pooling_size)
async_model = aegnn.asyncronous.make_model_asynchronous(model, radius, list(data_module.dims), edge_attr_func, log_flops=log_flops, log_runtime=log_runtime)

In [ ]:
async_mp_layer = async_model.pool5

In [ ]:
async_model.asy_pass_attribute('asy_pos', data.pos)

In [ ]:
x_f = data.x.clone()
x_f = elu(async_model.conv1(x_f, data.edge_index, data.edge_attr))
x_f = async_model.norm1(x_f)
x_f = elu(async_model.conv2(x_f, data.edge_index, data.edge_attr))
x_f = async_model.norm2(x_f)

x_sc = x_f.clone()
x_f = elu(async_model.conv3(x_f, data.edge_index, data.edge_attr))
x_f = async_model.norm3(x_f)
x_f = elu(async_model.conv4(x_f, data.edge_index, data.edge_attr))
x_f = async_model.norm4(x_f)
x_f = x_f + x_sc

x_f = elu(async_model.conv5(x_f, data.edge_index, data.edge_attr))
x_f = async_model.norm5(x_f)

In [ ]:
fm_to_pool = Data(x=x_f, pos=data.pos, batch=data.batch, edge_index=data.edge_index, edge_attr=data.edge_attr)

In [ ]:
print(f"Input graph with x = {fm_to_pool.x.shape} and pos = {fm_to_pool.pos.shape}")

Input graph with x = torch.Size([10000, 32]) and pos = torch.Size([10000, 2])


In [ ]:
graph_out = __dense_process(async_mp_layer, fm_to_pool.x, fm_to_pool.pos, edge_index=None)
graph_out #On obtient le coarse graphe comme dans le forward de la partie synchro

DataBatch(x=[29, 32], edge_index=[2, 142], pos=[29, 2], batch=[29], edge_attr=[142, 2], vdx=[29])

In [ ]:
async_mp_layer.asy_graph = Data(x=fm_to_pool.x, pos=fm_to_pool.pos) #On stock le graphe entier

In [ ]:
async_mp_layer.asy_graph.vdx = __get_clusters(async_mp_layer, pos=fm_to_pool.pos)  # voxel index of every node

In [ ]:
async_mp_layer.asy_graph.vdx

tensor([67, 43, 39,  ...,  1, 52, 15])

In [ ]:
a , index = torch.unique(async_mp_layer.asy_graph.vdx, sorted=True, return_inverse=True)

In [ ]:
index.shape

torch.Size([10000])

In [ ]:
_, argmax = scatter_max(fm_to_pool.x, index=index, dim=0)

In [ ]:
argmax.shape #Pour chaque chaque feature de chaque voxel node, on garde l'index de la node du graphe qui y a donné la valeur

torch.Size([29, 32])

In [ ]:
argmax_nodes = torch.unique(argmax)
async_mp_layer.asy_node_max_index = torch.flatten(argmax_nodes).long()
async_mp_layer.asy_node_max_index
#on stock les index des noeuds qui ont contribué à obtenir les features des voxel_nodes

tensor([   1,    3,    4,    7,    8,   10,   16,   21,   47,   51,   54,   55,
          59,   76,   86,   89,   96,  106,  111,  116,  125,  140,  141,  142,
         145,  147,  151,  159,  168,  173,  178,  183,  197,  200,  205,  209,
         211,  213,  220,  239,  240,  250,  253,  263,  277,  279,  295,  312,
         320,  322,  326,  331,  332,  342,  344,  350,  359,  380,  381,  385,
         392,  393,  399,  424,  433,  436,  448,  450,  451,  454,  457,  472,
         478,  481,  491,  501,  529,  541,  547,  551,  570,  593,  603,  617,
         627,  632,  667,  682,  688,  690,  696,  711,  714,  729,  764,  774,
         777,  783,  789,  792,  795,  804,  808,  815,  821,  825,  833,  838,
         841,  852,  863,  865,  867,  912,  920,  938,  949,  951,  961,  971,
         973,  985,  999, 1004, 1011, 1018, 1024, 1025, 1040, 1047, 1049, 1053,
        1062, 1070, 1078, 1089, 1090, 1111, 1128, 1140, 1145, 1155, 1160, 1175,
        1187, 1203, 1218, 1229, 1232, 12

In [ ]:
# On calcule la somme des positions pour chaque voxel
async_mp_layer.asy_voxel_pos_sum = scatter_sum(fm_to_pool.pos, index=index, dim=0)
#On stocke le nombre de nodes par voxel
async_mp_layer.asy_voxel_node_count = scatter_sum(torch.ones_like(async_mp_layer.asy_graph.vdx, device=fm_to_pool.x.device), index=index)


In [ ]:
async_mp_layer.asy_voxel_node_count

tensor([  1, 301,   4, 103, 638, 599, 710, 785, 334, 861, 295, 326, 322, 336,
        810, 275, 266, 287, 240, 434,  44,  40,  52,  92, 373, 643, 336, 266,
        227])

In [ ]:
async_mp_layer.asy_graph_coarse = graph_out.clone()
#Le graphe after pooling est aussi stocké. 

##### __graph processing

In [ ]:
model = aegnn.models.networks.GraphRes(data_module.name, input_shape, data_module.num_classes, pooling_size=pooling_size)
async_model = aegnn.asyncronous.make_model_asynchronous(model, radius, list(data_module.dims), edge_attr_func, log_flops=log_flops, log_runtime=log_runtime)
_ = async_model(data)

In [ ]:
async_model.asy_pass_attribute('asy_pos', event_new.pos)

In [ ]:
event_new.pos

tensor([[27., 16.]])

In [ ]:
async_model.asy_pass_attribute.listeners

[SplineConv(1, 8, dim=2),
 SplineConv(8, 16, dim=2),
 SplineConv(16, 16, dim=2),
 SplineConv(16, 16, dim=2),
 SplineConv(16, 32, dim=2),
 MaxPooling(voxel_size=[10, 10]),
 SplineConv(32, 32, dim=2),
 SplineConv(32, 32, dim=2),
 Linear(in_features=512, out_features=2, bias=False)]

In [ ]:
x_f = event_new.x.clone()
x_f = elu(async_model.conv1(x_f, event_new.edge_index, event_new.edge_attr))
x_f = async_model.norm1(x_f)
x_f = elu(async_model.conv2(x_f, event_new.edge_index, event_new.edge_attr))
x_f = async_model.norm2(x_f)

x_sc = x_f.clone()
x_f = elu(async_model.conv3(x_f, event_new.edge_index, event_new.edge_attr))
x_f = async_model.norm3(x_f)
x_f = elu(async_model.conv4(x_f, event_new.edge_index, event_new.edge_attr))
x_f = async_model.norm4(x_f)
x_f = x_f + x_sc

x_f = elu(async_model.conv5(x_f, event_new.edge_index, event_new.edge_attr))
x_f = async_model.norm5(x_f)

In [ ]:
async_mp_layer = async_model.pool5

In [ ]:
fm_to_pool = Data(x=x_f, pos=event_new.pos, batch=event_new.batch, edge_index=event_new.edge_index, edge_attr=event_new.edge_attr)

In [ ]:
pos = async_mp_layer.asy_pos
pos.shape

torch.Size([10001, 2])

In [ ]:
_, diff_idx = graph_changed_nodes(async_mp_layer, x= fm_to_pool.x)
_, new_idx = graph_new_nodes(async_mp_layer, x=fm_to_pool.x)

In [ ]:
print(f"Subgraph contains {new_idx.numel()} new and {diff_idx.numel()} diff nodes")

Subgraph contains 1 new and 409 diff nodes


In [ ]:
from aegnn.asyncronous.max_pool import __intersection

In [ ]:
#On détermine si les nodes qui ont changés ont contribué au coarse graph
replaced_idx = __intersection(diff_idx.long(), async_mp_layer.asy_node_max_index)
print(f"... from which {replaced_idx.numel()} nodes contributed to the coarse graph")

... from which 9 nodes contributed to the coarse graph


In [ ]:
torch.unique(replaced_idx, return_counts=True)

(tensor([ 108,  140,  475,  550,  801, 1155, 1227, 1726, 7969]),
 tensor([1, 1, 1, 1, 1, 1, 1, 1, 1]))

Si une node, qui a contribué à un voxel node, a changé, alors il faut reprendre toutes les nodes du voxel pour calculer à nouveau la voxel node.

In [ ]:
graph_vdx = getattr(async_mp_layer.asy_graph, "vdx")

In [ ]:
for idx in replaced_idx:
    nodes_voxel = torch.nonzero(torch.eq(graph_vdx, graph_vdx[idx]))[:, 0]
    #On détermine les nodes du voxel auquel appartient la node contributrice qui a changé de features
    diff_idx = torch.cat([diff_idx, nodes_voxel])

    voxel_idx = async_mp_layer.asy_graph.vdx[idx]
    coarse_idx = torch.eq(getattr(async_mp_layer.asy_graph_coarse, "vdx"), voxel_idx)

    async_mp_layer.asy_graph_coarse.x[coarse_idx] = -999999 
    async_mp_layer.asy_voxel_pos_sum[coarse_idx] -= pos[idx]
    async_mp_layer.asy_voxel_node_count[coarse_idx] -= 1


In [ ]:
diff_idx.shape

torch.Size([6415])

In [ ]:
#On récupère les features et positions des nodes qui ont changé 
update_idx = torch.cat([diff_idx, new_idx])
x_update = x_f[update_idx, :]
pos_update = pos[update_idx, :]

In [ ]:
update_idx.shape

torch.Size([6416])

In [ ]:
vdx_update = __get_clusters(async_mp_layer, pos=pos_update)

In [ ]:
async_mp_layer.asy_graph_coarse.x.shape
x_update.shape

torch.Size([6416, 32])

In [ ]:
#On concatène les update avec les variables en mémoire 
x_scatter = torch.cat([x_update, async_mp_layer.asy_graph_coarse.x])
node_index_scatter = torch.cat([update_idx, async_mp_layer.asy_node_max_index])
pos_sum_scatter = torch.cat([pos_update, async_mp_layer.asy_voxel_pos_sum])
node_count_scatter = torch.cat([torch.ones_like(update_idx, device=x_f.device), async_mp_layer.asy_voxel_node_count])
clusters = torch.cat([vdx_update, getattr(async_mp_layer.asy_graph_coarse, "vdx")])

In [ ]:
#On calcule le max sur chaque cluster 
clusters_unique, index = torch.unique(clusters, sorted=True, return_inverse=True)
x_max, argmax = scatter_max(x_scatter, index=index, dim=0)

In [ ]:
torch.unique(argmax, return_counts=True)

(tensor([   1,    2,   18,   22,   25,   53,   85,  312,  421,  436,  460,  463,
          471,  472,  519,  521,  546,  561,  872,  941, 6092, 6101, 6111, 6126,
         6130, 6141, 6142, 6152, 6172, 6177, 6178, 6179, 6185, 6193, 6210, 6269,
         6320, 6358, 6416, 6417, 6418, 6419, 6420, 6421, 6423, 6424, 6425, 6426,
         6428, 6429, 6430, 6431, 6432, 6433, 6434, 6435, 6436, 6437, 6438, 6439,
         6440, 6441, 6442, 6443, 6444]),
 tensor([ 1,  1,  1,  1,  1,  1,  1,  2,  1,  1,  1,  1,  4,  2,  2,  1,  4,  1,
          5,  2,  2,  2,  1,  1,  2,  1,  2,  2,  4,  1,  1,  1,  1,  1,  1,  2,
          1,  4, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32,
         32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32]))

In [ ]:
#On calcule la position des voxels nodes
voxel_pos_sum = scatter_sum(pos_sum_scatter, index=index, dim=0)
voxel_node_count = scatter_sum(node_count_scatter, index=index)
pos_mean = torch.div(voxel_pos_sum, voxel_node_count.view(-1, 1))

In [ ]:
vdx = torch.cat([getattr(async_mp_layer.asy_graph, "vdx"), vdx_update[diff_idx.numel():]]) # On ajoute la nouvelle node index
vdx[diff_idx] = vdx_update[:diff_idx.numel()] #On change les diff index

In [ ]:
from torch_geometric.nn.pool.pool import pool_edge


In [ ]:
edges_coarse = torch.empty((2, 0), dtype=torch.long, device=x_f.device)

Il n'y a pas d'edge_index dans event new.

On recréé donc un coarse graph et on update le graphe avant coarsening.

In [ ]:
graph_out = Data(x=x_max, pos=pos_mean, edge_index=edges_coarse)
async_mp_layer.asy_graph = Data(x=x_f, pos=pos, edge_index=edge_index, vdx=vdx).clone()

In [ ]:
async_mp_layer.asy_node_max_index = torch.flatten(node_index_scatter[argmax]).long()
async_mp_layer.asy_voxel_pos_sum = voxel_pos_sum
async_mp_layer.asy_voxel_node_count = voxel_node_count
async_mp_layer.asy_graph_coarse = Data(x=graph_out.x, pos=graph_out.pos, edge_index=edges_coarse, vdx=clusters_unique)


In [ ]:
#On considère dans l'asyncronous processing que les event sont dans le même batch 
graph_out.batch = torch.zeros(graph_out.num_nodes, dtype=torch.long, device=graph_out.x.device)

In [ ]:
#On assigne les positions du coarse graph aux autres 
async_mp_layer.asy_pass_attribute('asy_pos', graph_out.pos)

In [ ]:
graph_out.pos.shape

torch.Size([29, 2])

# Test

In [ ]:
pl.seed_everything(12345, workers=True)

Seed set to 12345


12345

In [ ]:
device = "cuda" if torch.cuda.is_available() else "cpu"
device

'cuda'

In [ ]:
model_path = "/home/ibrahim/event_graph_aegnn/external/training_log_aegnn_multiseed/checkpoints/ncars/recognition/20260619120358/best-24-0.0000.ckpt"

In [ ]:
data_module = aegnn.datasets.NCars(batch_size=16, shuffle=False, )
data_module.setup()

In [ ]:
test_loader = data_module.test_dataloader()
val_loader = data_module.val_dataloader()

In [ ]:

batch_init = next(iter(test_loader))

batch = batch_init.clone()
print(batch)

batch_idx = batch.batch
batch.x
num_graphs = len(torch.unique(batch.batch))
for graph_idx in range(num_graphs):
    t_min = torch.min(batch.x[batch.batch == graph_idx][:,2])
    t_max = torch.max(batch.x[batch.batch == graph_idx][:,2])
    print(f"Before: {t_max} | {t_min}")
    
    mask = batch.batch == graph_idx
    batch.x[mask, 2] = (batch.x[mask, 2]-t_min)/(t_max-t_min)
    
    print(batch.x[batch.batch == graph_idx][:,2].shape)

    t_min = torch.min(batch.x[batch.batch == graph_idx][:,2])
    t_max = torch.max(batch.x[batch.batch == graph_idx][:,2])
    print(f"After: {t_max} | {t_min}")

# Verification
# for graph_idx in range(num_graphs):

#     print("\n min: ",torch.min(batch.x[batch.batch == graph_idx][:,2]))

#     print("\n max: ",torch.max(batch.x[batch.batch == graph_idx][:,2]))

DataBatch(x=[60953, 4], pos=[60953, 3], file_id=[16], label=[16], y=[16], edge_index=[2, 1683194], edge_attr=[1683194, 3], batch=[60953], ptr=[17])
Before: 0.09984099864959717 | 0.0
torch.Size([1634])
After: 1.0 | 0.0
Before: 0.09998899698257446 | 0.0
torch.Size([5283])
After: 1.0 | 0.0
Before: 0.09982900321483612 | 0.0
torch.Size([739])
After: 1.0 | 0.0
Before: 0.09983900189399719 | 0.0
torch.Size([560])
After: 1.0 | 0.0
Before: 0.09994400292634964 | 0.0
torch.Size([4024])
After: 1.0 | 0.0
Before: 0.09995699673891068 | 0.0
torch.Size([3941])
After: 1.0 | 0.0
Before: 0.09987399727106094 | 0.0
torch.Size([1360])
After: 1.0 | 0.0
Before: 0.09997399896383286 | 0.0
torch.Size([10000])
After: 1.0 | 0.0
Before: 0.09997600317001343 | 0.0
torch.Size([5842])
After: 1.0 | 0.0
Before: 0.09996800124645233 | 0.0
torch.Size([5271])
After: 1.0 | 0.0
Before: 0.09977299720048904 | 0.0
torch.Size([1576])
After: 1.0 | 0.0
Before: 0.0999630019068718 | 0.0
torch.Size([4370])
After: 1.0 | 0.0
Before: 0.0998

In [ ]:
input_shape = torch.tensor([*data_module.dims, batch.pos.shape[-1]], device='cpu')
pooling_size = (10,10)
input_shape

NameError: name 'batch' is not defined

In [ ]:
input_shape[:2] - 1

tensor([119,  99])

In [ ]:
#model_path = "/home/ibrahim/event_graph_aegnn/training_log_aegnn/checkpoints/ncars/recognition/20260607163336_xytp/best-23-0.0000.ckpt"
#model_path = "/home/ibrahim/event_graph_aegnn/training_log_aegnn/checkpoints/ncars/recognition/20260608071926/best-27-0.0000_xytp.ckpt"
#model_path = "/home/ibrahim/event_graph_aegnn/training_log_aegnn/checkpoints/ncars/recognition/20260609211206/best-29-0.0000.ckpt" #xytp
#model_path = "/home/ibrahim/event_graph_aegnn/training_log_aegnn/checkpoints/ncars/recognition/20260609071610/best-18-0.0000.ckpt" #xy
# model_path = "/home/ibrahim/event_graph_aegnn/training_log_aegnn/checkpoints/ncars/recognition/20260610152918/best-21-0.0000.ckpt" #xytp normalized
# model_path = "/home/ibrahim/event_graph_aegnn/external/aegnn/training_log_aegnn/checkpoints/ncars/recognition/20260611111055/best-23-0.0000.ckpt"
# model_path = "/home/ibrahim/event_graph_aegnn/training_log_aegnn/checkpoints/ncars/recognition/20260611133107/best-26-0.0000.ckpt"
#model_path = "/home/ibrahim/event_graph_aegnn/training_log_aegnn/checkpoints/ncars/recognition/20260612000137/best-24-0.0000.ckpt"
#model_path = "/home/ibrahim/event_graph_aegnn/external/training_log_aegnn_multiseed/checkpoints/ncars/recognition/20260619120358/best-24-0.0000.ckpt"
#model_path = "/home/ibrahim/event_graph_aegnn/external/training_log_aegnn_multiseed/checkpoints/ncars/recognition/20260619141409/best-08-0.0000.ckpt"
# model_path = "/home/ibrahim/event_graph_aegnn/external/training_log_aegnn_multiseed/checkpoints/ncars/recognition/20260626232410/best-19-0.0000.ckpt"
# model_path = "/home/ibrahim/event_graph_aegnn/external/training_log_aegnn_multiseed/checkpoints/ncars/recognition/20260627150531/last.ckpt"
# model_path = "/home/ibrahim/event_graph_aegnn/external/training_log_aegnn_multiseed/checkpoints/ncars/recognition/20260627091942/last.ckpt"
# model_path = "/home/ibrahim/event_graph_aegnn/external/training_log_aegnn_multiseed/checkpoints/ncars/recognition/20260626232410/last.ckpt"
# model_path = "/home/ibrahim/event_graph_aegnn/external/training_log_aegnn_multiseed/checkpoints/ncars/recognition/20260627225409/last.ckpt"
# model_path = "/home/ibrahim/event_graph_aegnn/external/training_log_aegnn_multiseed/checkpoints/ncars/recognition/20260629094433/best-07-0.0000.ckpt"
model_path = "/home/ibrahim/event_graph_aegnn/external/training_log_aegnn_multiseed/checkpoints/ncars/recognition/20260629121937/best-00-0.0000.ckpt"
model = aegnn.models.recognition.RecognitionModel.load_from_checkpoint(model_path,
                                                                        network='graph_res', dataset='ncars', num_classes = data_module.num_classes, img_shape=data_module.dims, bias = True,
                                      root_weight = True )

RuntimeError: Error(s) in loading state_dict for RecognitionModel:
	size mismatch for model.conv1.weight: copying a param with shape torch.Size([8, 3, 8]) from checkpoint, the shape in current model is torch.Size([8, 1, 8]).
	size mismatch for model.conv1.lin.weight: copying a param with shape torch.Size([8, 3]) from checkpoint, the shape in current model is torch.Size([8, 1]).

In [ ]:
tmp_model = aegnn.models.recognition.RecognitionModel(
    network="graph_res",
    dataset="ncars",  # pas ncars si ton checkpoint vient de NCaltech
    num_classes=data_module.num_classes,
    img_shape=data_module.dims,
    bias = True,
    root_weight = True, 
)

for k in list(tmp_model.state_dict().keys())[:30]:
    print(k)

In [ ]:
# import torch
# model_path = "/home/ibrahim/event_graph_aegnn/training_log_aegnn/checkpoints/ncars/recognition/20260607163336_xytp/best-23-0.0000.ckpt"
# ckpt = torch.load(model_path, map_location="cpu")
# print(ckpt.keys())
# print(ckpt.get("hyper_parameters", None))
# print(list(ckpt["state_dict"].keys())[:20])

dict_keys(['epoch', 'global_step', 'pytorch-lightning_version', 'state_dict', 'loops', 'callbacks', 'optimizer_states', 'lr_schedulers', 'datamodule_hparams_name', 'datamodule_hyper_parameters'])
None
['model.conv1.weight', 'model.conv1.bias', 'model.conv1.kernel_size', 'model.conv1.is_open_spline', 'model.conv1.lin.weight', 'model.norm1.module.weight', 'model.norm1.module.bias', 'model.norm1.module.running_mean', 'model.norm1.module.running_var', 'model.norm1.module.num_batches_tracked', 'model.conv2.weight', 'model.conv2.bias', 'model.conv2.kernel_size', 'model.conv2.is_open_spline', 'model.conv2.lin.weight', 'model.norm2.module.weight', 'model.norm2.module.bias', 'model.norm2.module.running_mean', 'model.norm2.module.running_var', 'model.norm2.module.num_batches_tracked']


In [ ]:
ckpt_keys = set(ckpt["state_dict"].keys())
model_keys = set(tmp_model.state_dict().keys())

print("Unexpected in ckpt:")
for k in sorted(ckpt_keys - model_keys):
    print(k)

print("Missing from ckpt:")
for k in sorted(model_keys - ckpt_keys):
    print(k)

Unexpected in ckpt:
Missing from ckpt:


In [ ]:
batch = batch.to(device)

In [ ]:
out = model(batch)

In [ ]:
out.shape

torch.Size([64, 2])

In [ ]:
import lightning.pytorch as pl

trainer = pl.Trainer()

GPU available: True (cuda), used: True
TPU available: False, using: 0 TPU cores
💡 Tip: For seamless cloud logging and experiment tracking, try installing [litlogger](https://pypi.org/project/litlogger/) to enable LitLogger, which logs metrics and artifacts automatically to the Lightning Experiments platform.


In [ ]:
len(data_module.test_dataset)

8606

In [ ]:
test_loader = data_module.test_dataloader()

In [ ]:
results = trainer.test(model, dataloaders=test_loader)

💡 Tip: For seamless cloud uploads and versioning, try installing [litmodels](https://pypi.org/project/litmodels/) to enable LitModelCheckpoint, which syncs automatically with the Lightning model registry.
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]


Testing: |          | 0/? [00:00<?, ?it/s]


 **************file id ['sequence_1961', 'sequence_5556', 'sequence_6639', 'sequence_7203', 'sequence_762', 'sequence_6426', 'sequence_1168', 'sequence_1384', 'sequence_6007', 'sequence_4712', 'sequence_3699', 'sequence_2831', 'sequence_8586', 'sequence_6217', 'sequence_5066', 'sequence_5029']*********

init X_max: tensor(1., device='cuda:0')
init mean: tensor(0.4227, device='cuda:0')
init X_var min: tensor(0.2440, device='cuda:0')


conv1 max: tensor(0.9031, device='cuda:0')
conv1 mean: tensor(0.1240, device='cuda:0')
conv1 var : tensor(0.0984, device='cuda:0')


conv2 max: tensor(4.3191, device='cuda:0')
conv2 mean: tensor(0.1533, device='cuda:0')
conv2 var min: tensor(0.7110, device='cuda:0')


conv3 max: tensor(5.4362, device='cuda:0')
conv3 mean: tensor(0.1376, device='cuda:0')
conv3 var : tensor(0.9568, device='cuda:0')


conv4 max: tensor(9.9926, device='cuda:0')
conv4 mean: tensor(0.3507, device='cuda:0')
conv4 var: tensor(1.4232, device='cuda:0')


conv5 max: tensor(7.1969, d

In [ ]:
results

[{'Test/loss': 0.7222890257835388,
  'Test/Accuracy': 0.8167557716369629,
  'Test/Accuracy_Top1': 0.8167557716369629}]

In [ ]:
trainer.test(model, dataloaders=val_loader)

LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]


Testing: |          | 0/? [00:00<?, ?it/s]


 **************file id ['sequence_13373', 'sequence_14055', 'sequence_14593', 'sequence_14548', 'sequence_14321', 'sequence_15369', 'sequence_13706', 'sequence_14938', 'sequence_13808', 'sequence_14563', 'sequence_13324', 'sequence_15007', 'sequence_14118', 'sequence_13952', 'sequence_12982', 'sequence_14617']*********

init X_max: tensor(1., device='cuda:0')
init mean: tensor(0.5105, device='cuda:0')
init X_var min: tensor(0.2499, device='cuda:0')


conv1 max: tensor(0.9122, device='cuda:0')
conv1 mean: tensor(0.1441, device='cuda:0')
conv1 var : tensor(0.1143, device='cuda:0')


conv2 max: tensor(4.3583, device='cuda:0')
conv2 mean: tensor(0.1676, device='cuda:0')
conv2 var min: tensor(0.7348, device='cuda:0')


conv3 max: tensor(5.5989, device='cuda:0')
conv3 mean: tensor(0.1408, device='cuda:0')
conv3 var : tensor(0.9729, device='cuda:0')


conv4 max: tensor(11.1791, device='cuda:0')
conv4 mean: tensor(0.3354, device='cuda:0')
conv4 var: tensor(1.4440, device='cuda:0')


conv5 max

[{'Test/loss': 0.2912876605987549,
  'Test/Accuracy': 0.9008935689926147,
  'Test/Accuracy_Top1': 0.9008935689926147}]

# Debug 

In [ ]:
data_module = aegnn.datasets.NCars(batch_size=16, shuffle=False, )
data_module.setup()

In [ ]:
val_loader = data_module.val_dataloader()
test_loader = data_module.test_dataloader()

In [ ]:
trainer.test(model, dataloaders=test_loader)

LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]


Testing: |          | 0/? [00:00<?, ?it/s]

tensor([ 0,  0,  0,  ..., 15, 15, 15], device='cuda:0')
tensor([ 0,  0,  0,  ..., 15, 15, 15], device='cuda:0')
tensor([ 0,  0,  0,  ..., 15, 15, 15], device='cuda:0')
tensor([ 0,  0,  0,  ..., 15, 15, 15], device='cuda:0')
tensor([ 0,  0,  0,  ..., 15, 15, 15], device='cuda:0')
tensor([ 0,  0,  0,  ..., 15, 15, 15], device='cuda:0')
tensor([ 0,  0,  0,  ..., 15, 15, 15], device='cuda:0')
tensor([ 0,  0,  0,  ..., 15, 15, 15], device='cuda:0')
tensor([ 0,  0,  0,  ..., 15, 15, 15], device='cuda:0')
tensor([ 0,  0,  0,  ..., 15, 15, 15], device='cuda:0')
tensor([ 0,  0,  0,  ..., 15, 15, 15], device='cuda:0')
tensor([ 0,  0,  0,  ..., 15, 15, 15], device='cuda:0')
tensor([ 0,  0,  0,  ..., 15, 15, 15], device='cuda:0')
tensor([ 0,  0,  0,  ..., 15, 15, 15], device='cuda:0')
tensor([ 0,  0,  0,  ..., 15, 15, 15], device='cuda:0')
tensor([ 0,  0,  0,  ..., 15, 15, 15], device='cuda:0')
tensor([ 0,  0,  0,  ..., 15, 15, 15], device='cuda:0')
tensor([ 0,  0,  0,  ..., 15, 15, 15], device='c

[{'Test/loss': 0.43036356568336487,
  'Test/Accuracy': 0.8618406057357788,
  'Test/Accuracy_Top1': 0.8618406057357788}]

In [ ]:
device = "cuda" if torch.cuda.is_available() else "cpu"
device

'cuda'

In [ ]:
import torch.nn as nn
from torchmetrics.functional import accuracy

criterion = nn.CrossEntropyLoss()


In [ ]:
model.model.to(device)

GraphRes(
  (conv1): SplineConv(1, 8, dim=3)
  (norm1): BatchNorm(8, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (conv2): SplineConv(8, 16, dim=3)
  (norm2): BatchNorm(16, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (conv3): SplineConv(16, 16, dim=3)
  (norm3): BatchNorm(16, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (conv4): SplineConv(16, 16, dim=3)
  (norm4): BatchNorm(16, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (conv5): SplineConv(16, 32, dim=3)
  (norm5): BatchNorm(32, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (pool5): MaxPooling(voxel_size=[16, 12])
  (conv6): SplineConv(32, 32, dim=3)
  (norm6): BatchNorm(32, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (conv7): SplineConv(32, 32, dim=3)
  (norm7): BatchNorm(32, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (pool7): MaxPoolingX(voxel_size=tensor([30, 25]), size=16)
  (fc): 

In [ ]:
model.model(batch)

In [ ]:
val_data = data_module.val_dataset[0]
test_data = data_module.test_dataset[0]

for name, d in [("val", val_data), ("test", test_data)]:
    print("\n", name)
    print(d)
    print("x:", d.x.shape, d.x.dtype, d.x.device)
    print("pos:", d.pos.shape, d.pos.dtype, d.pos.device)
    print("edge_index:", d.edge_index.shape, d.edge_index.dtype, d.edge_index.device)
    print("edge_attr:", d.edge_attr.shape if hasattr(d, "edge_attr") else None,
          d.edge_attr.dtype if hasattr(d, "edge_attr") else None,
          d.edge_attr.device if hasattr(d, "edge_attr") else None)
    print("edge_index max:", d.edge_index.max().item())
    print("num_nodes:", d.num_nodes)


 val
Data(x=[9117, 1], pos=[9117, 3], file_id='sequence_13859', label=[1], y=[1], edge_index=[2, 277957], edge_attr=[277957, 3])
x: torch.Size([9117, 1]) torch.float32 cpu
pos: torch.Size([9117, 3]) torch.float32 cpu
edge_index: torch.Size([2, 277957]) torch.int64 cpu
edge_attr: torch.Size([277957, 3]) torch.float32 cpu
edge_index max: 9116
num_nodes: 9117

 test
Data(x=[3902, 1], pos=[3902, 3], file_id='sequence_7264', label=[1], y=[1], edge_index=[2, 105898], edge_attr=[105898, 3])
x: torch.Size([3902, 1]) torch.float32 cpu
pos: torch.Size([3902, 3]) torch.float32 cpu
edge_index: torch.Size([2, 105898]) torch.int64 cpu
edge_attr: torch.Size([105898, 3]) torch.float32 cpu
edge_index max: 3901
num_nodes: 3902


In [ ]:
assert val_data.edge_index.max() < val_data.num_nodes
assert test_data.edge_index.max() < test_data.num_nodes
assert val_data.edge_attr.size(-1) == test_data.edge_attr.size(-1)

In [ ]:
print(torch.__version__)
print(torch_geometric.__version__)
print(torch_spline_conv.__version__)
print(model.model.conv1.kernel_size.device)

2.7.1+cu128
2.7.0
1.2.2+pt27cu128
cuda:0


In [ ]:
val_acc = 0
val_loss = 0
n = 0

model.model.eval()
model.model.to(device)

with torch.no_grad():
    for batch in val_loader:
        
        batch = batch.to(device)

        out = model.model(batch)
        print(f"out : {out.shape}, y: {batch.y.shape}")
        loss = criterion(out, batch.y)

        val_loss += loss.item()

        y_prediction = torch.argmax(out, dim=-1)
        val_acc += accuracy(preds=y_prediction, target=batch.y, task="multiclass", num_classes=data_module.num_classes)

        n += 1
        print(f"okay! {n}")

val_acc /= n
val_loss /= n

/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:365: operator(): block: [10,0,0], thread: [96,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:365: operator(): block: [10,0,0], thread: [97,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:365: operator(): block: [10,0,0], thread: [98,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:365: operator(): block: [10,0,0], thread: [99,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:365: operator(): block: [10,0,0], thread: [100,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/ScatterGa

RuntimeError: CUDA error: CUBLAS_STATUS_EXECUTION_FAILED when calling cublasLtMatmul with transpose_mat1 1 transpose_mat2 0 m 2 n 16 k 512 mat1_ld 512 mat2_ld 512 result_ld 2 abcType 0 computeType 68 scaleType 0

In [ ]:
val_loss

0.13287358857595769

In [ ]:
val_acc

tensor(0.9555, device='cuda:0')

In [ ]:
batch

DataBatch(x=[1353, 1], pos=[1353, 3], file_id=[1], label=[1], y=[1], edge_index=[2, 42762], edge_attr=[42762, 3], batch=[1353], ptr=[2])

In [ ]:
dataset=data_module.val_dataset

In [ ]:
from torch_geometric.data import Batch

model.model.eval()

i, j = 0, 1

with torch.no_grad():
    out_single = model.model(Batch.from_data_list([dataset[i]]).to(device))

    out_pair = model.model(Batch.from_data_list([dataset[i], dataset[j]]).to(device))[0:1]

print("max diff:", (out_single - out_pair).abs().max().item())
print("out single:", out_single)
print("out pair:", out_pair)

pool7 x shape: torch.Size([16, 32])
data.num_graphs: 1
pooled batch unique: tensor([0], device='cuda:0')
pooled batch bincount: tensor([25], device='cuda:0')
pool7 x shape: torch.Size([32, 32])
data.num_graphs: 2
pooled batch unique: tensor([0, 1], device='cuda:0')
pooled batch bincount: tensor([25, 12], device='cuda:0')
max diff: 1.6648247241973877
out single: tensor([[ 4.1193, -4.9140]], device='cuda:0')
out pair: tensor([[ 2.4545, -4.3127]], device='cuda:0')


In [ ]:
from torch_geometric.data import Batch

model.model.eval()

i, j = 0, 1

with torch.no_grad():
    out_single = model.model(Batch.from_data_list([dataset[i]]).to(device))
    print("out single1:", out_single)
    out_single = model.model(Batch.from_data_list([dataset[i]]).to(device))
    out_pair = model.model(Batch.from_data_list([dataset[i], dataset[j]]).to(device))[0:1]
    print("out single2:", out_single)

print("max diff:", (out_single - out_pair).abs().max().item())
print("out single:", out_single)
print("out pair:", out_pair)

pool7 x shape: torch.Size([16, 32])
data.num_graphs: 1
pooled batch unique: tensor([0], device='cuda:0')
pooled batch bincount: tensor([25], device='cuda:0')
out single1: tensor([[ 4.1193, -4.9140]], device='cuda:0')
pool7 x shape: torch.Size([16, 32])
data.num_graphs: 1
pooled batch unique: tensor([0], device='cuda:0')
pooled batch bincount: tensor([25], device='cuda:0')
pool7 x shape: torch.Size([32, 32])
data.num_graphs: 2
pooled batch unique: tensor([0, 1], device='cuda:0')
pooled batch bincount: tensor([25, 12], device='cuda:0')
out single2: tensor([[ 4.1193, -4.9140]], device='cuda:0')
max diff: 1.6648247241973877
out single: tensor([[ 4.1193, -4.9140]], device='cuda:0')
out pair: tensor([[ 2.4545, -4.3127]], device='cuda:0')


In [ ]:
import torch
import torch_geometric

from torch.nn import Linear
from torch.nn.functional import elu
from torch_geometric.nn.conv import SplineConv
from torch_geometric.nn.norm import BatchNorm
from torch_geometric.transforms import Cartesian
from torch_geometric.data import Data,Batch


from aegnn.models.layer import MaxPooling, MaxPoolingX

In [ ]:
device = "cuda" if torch.cuda.is_available() else "cpu"
device

'cuda'

In [ ]:
kernel_size = 2
n = [1, 8, 16, 16, 16, 32, 32, 32, 32]
pooling_outputs = 32
pooling_size = (10,10)

In [ ]:
bias = True
root_weight =True
dim = 3
num_outputs = 2
input_shape = torch.Tensor([120, 100,   3])

In [ ]:
conv1 = SplineConv(n[0], n[1], dim=dim, kernel_size=kernel_size, bias=bias, root_weight=root_weight)
norm1 = BatchNorm(in_channels=n[1])
conv2 = SplineConv(n[1], n[2], dim=dim, kernel_size=kernel_size, bias=bias, root_weight=root_weight)
norm2 = BatchNorm(in_channels=n[2])

conv3 = SplineConv(n[2], n[3], dim=dim, kernel_size=kernel_size, bias=bias, root_weight=root_weight)
norm3 = BatchNorm(in_channels=n[3])
conv4 = SplineConv(n[3], n[4], dim=dim, kernel_size=kernel_size, bias=bias, root_weight=root_weight)
norm4 = BatchNorm(in_channels=n[4])

conv5 = SplineConv(n[4], n[5], dim=dim, kernel_size=kernel_size, bias=bias, root_weight=root_weight)
norm5 = BatchNorm(in_channels=n[5])
pool5 = MaxPooling(pooling_size, transform=Cartesian(norm=True, cat=False))

conv6 = SplineConv(n[5], n[6], dim=dim, kernel_size=kernel_size, bias=bias, root_weight=root_weight)
norm6 = BatchNorm(in_channels=n[6])
conv7 = SplineConv(n[6], n[7], dim=dim, kernel_size=kernel_size, bias=bias, root_weight=root_weight)
norm7 = BatchNorm(in_channels=n[7])

pool7 = MaxPoolingX(input_shape[:2] // 4, size=16,  start=[0., 0.], end=input_shape[:2] )
fc = Linear(pooling_outputs * 16, out_features=num_outputs, bias=bias)

In [ ]:
data_module = aegnn.datasets.NCars(batch_size=64, shuffle=False, )
data_module.setup()
dataset=data_module.test_dataset

In [ ]:
batch_single = Batch.from_data_list([dataset[0]]).to(device)
batch_pair = Batch.from_data_list([dataset[0], dataset[1]]).to(device)

RuntimeError: CUDA error: device-side assert triggered
CUDA kernel errors might be asynchronously reported at some other API call, so the stacktrace below might be incorrect.
For debugging consider passing CUDA_LAUNCH_BLOCKING=1
Compile with `TORCH_USE_CUDA_DSA` to enable device-side assertions.


In [ ]:
def compare(name, a, b, batch_pair):
    b0 = b[batch_pair.batch == 0]
    print(name)
    print("shapes:", a.shape, b0.shape)
    print("max diff:", (a - b0).abs().max().item())
    print("mean diff:", (a - b0).abs().mean().item())

### Forward

#### Conv1

In [ ]:
conv1 =conv1.to(device)
conv1.eval()


SplineConv(1, 8, dim=3)

In [ ]:
x1 = elu(conv1(batch_single.x, batch_single.edge_index, batch_single.edge_attr))
x2 = elu(conv1(batch_pair.x, batch_pair.edge_index, batch_pair.edge_attr))
compare("conv1", x1, x2, batch_pair)

conv1
shapes: torch.Size([3902, 8]) torch.Size([3902, 8])
max diff: 5.960464477539063e-08
mean diff: 1.2300547247434679e-09


#### Norm1

In [ ]:
norm1.eval()
norm1.to(device)

BatchNorm(8, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)

In [ ]:
x1 = norm1(x1)
x2 = norm1(x2)
compare("norm1", x1, x2, batch_pair)

norm1
shapes: torch.Size([3902, 8]) torch.Size([3902, 8])
max diff: 5.960464477539063e-08
mean diff: 1.2300547247434679e-09


#### Conv2

In [ ]:
conv2.to(device)
conv2.eval()


SplineConv(8, 16, dim=3)

In [ ]:
x1 = elu(conv2(x1, batch_single.edge_index, batch_single.edge_attr))
x2 = elu(conv2(x2, batch_pair.edge_index, batch_pair.edge_attr))
compare("conv2", x1, x2, batch_pair)

conv2
shapes: torch.Size([3902, 16]) torch.Size([3902, 16])
max diff: 5.960464477539063e-08
mean diff: 3.904342182892151e-09


#### Norm2

In [ ]:
norm2.eval()
norm2.to(device)

BatchNorm(16, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)

In [ ]:
x1 = norm2(x1)
x2 = norm2(x2)
compare("norm2", x1, x2, batch_pair)

norm2
shapes: torch.Size([3902, 16]) torch.Size([3902, 16])
max diff: 5.960464477539063e-08
mean diff: 3.904342182892151e-09


In [ ]:
x_sc1 = x1.clone()
x_sc2 = x2.clone()

#### Conv3

In [ ]:
conv3.to(device)
conv3.eval()


SplineConv(16, 16, dim=3)

In [ ]:
x1 = elu(conv3(x1, batch_single.edge_index, batch_single.edge_attr))
x2 = elu(conv3(x2, batch_pair.edge_index, batch_pair.edge_attr))
compare("conv3", x1, x2, batch_pair)

conv3
shapes: torch.Size([3902, 16]) torch.Size([3902, 16])
max diff: 1.1920928955078125e-07
mean diff: 5.413789860853058e-09


#### Norm3

In [ ]:
norm3.eval()
norm3.to(device)

BatchNorm(16, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)

In [ ]:
x1 = norm3(x1)
x2 = norm3(x2)
compare("norm3", x1, x2, batch_pair)

norm3
shapes: torch.Size([3902, 16]) torch.Size([3902, 16])
max diff: 1.1920928955078125e-07
mean diff: 5.413789860853058e-09


#### Conv4

In [ ]:
conv4.to(device)
conv4.eval()


SplineConv(16, 16, dim=3)

In [ ]:
x1 = elu(conv4(x1, batch_single.edge_index, batch_single.edge_attr))
x2 = elu(conv4(x2, batch_pair.edge_index, batch_pair.edge_attr))
compare("conv4", x1, x2, batch_pair)

conv4
shapes: torch.Size([3902, 16]) torch.Size([3902, 16])
max diff: 5.960464477539063e-08
mean diff: 4.141057718953789e-09


#### Norm4

In [ ]:
norm4.eval()
norm4.to(device)

BatchNorm(16, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)

In [ ]:
x1 = norm4(x1)
x2 = norm4(x2)
compare("norm4", x1, x2, batch_pair)

norm4
shapes: torch.Size([3902, 16]) torch.Size([3902, 16])
max diff: 5.960464477539063e-08
mean diff: 4.141057274864579e-09


In [ ]:
x1 = x1+x_sc1
x2 = x2+x_sc2

#### Conv5

In [ ]:
conv5.to(device)
conv5.eval()


SplineConv(16, 32, dim=3)

In [ ]:
x1 = elu(conv5(x1, batch_single.edge_index, batch_single.edge_attr))
x2 = elu(conv5(x2, batch_pair.edge_index, batch_pair.edge_attr))
compare("conv5", x1, x2, batch_pair)

conv5
shapes: torch.Size([3902, 32]) torch.Size([3902, 32])
max diff: 9.685754776000977e-08
mean diff: 6.129769136720142e-09


#### Norm5

In [ ]:
norm5.eval()
norm5.to(device)

BatchNorm(32, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)

In [ ]:
x1 = norm5(x1)
x2 = norm5(x2)
compare("norm5", x1, x2, batch_pair)

norm5
shapes: torch.Size([3902, 32]) torch.Size([3902, 32])
max diff: 9.685754776000977e-08
mean diff: 6.129760699025155e-09


#### Pooling

In [ ]:
batch_single = pool5(x1, pos=batch_single.pos, batch=batch_single.batch, edge_index=batch_single.edge_index, return_data_obj=True)
batch_pair = pool5(x2, pos=batch_pair.pos, batch=batch_pair.batch, edge_index=batch_pair.edge_index, return_data_obj=True)

x1 = batch_single.x.clone()
x2 = batch_pair.x.clone()


compare("pool5", x1, x2, batch_pair)

pool5
shapes: torch.Size([48, 32]) torch.Size([48, 32])
max diff: 5.960464477539063e-08
mean diff: 7.282442648914866e-09


In [ ]:
x_sc1 = x1.clone()
x_sc2 = x2.clone()

#### Conv6

In [ ]:
conv6.to(device)
conv6.eval()


SplineConv(32, 32, dim=3)

In [ ]:
x1 = elu(conv6(x1, batch_single.edge_index, batch_single.edge_attr))
x2 = elu(conv6(x2, batch_pair.edge_index, batch_pair.edge_attr))
compare("conv6", x1, x2, batch_pair)

conv6
shapes: torch.Size([48, 32]) torch.Size([48, 32])
max diff: 4.470348358154297e-08
mean diff: 6.737006508217291e-09


#### Norm6

In [ ]:
norm6.eval()
norm6.to(device)

BatchNorm(32, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)

In [ ]:
x1 = norm6(x1)
x2 = norm6(x2)
compare("norm6", x1, x2, batch_pair)

norm6
shapes: torch.Size([48, 32]) torch.Size([48, 32])
max diff: 4.470348358154297e-08
mean diff: 6.737006508217291e-09


#### Conv7

In [ ]:
conv7.to(device)
conv7.eval()


SplineConv(32, 32, dim=3)

In [ ]:
x1 = elu(conv7(x1, batch_single.edge_index, batch_single.edge_attr))
x2 = elu(conv7(x2, batch_pair.edge_index, batch_pair.edge_attr))
compare("conv7", x1, x2, batch_pair)

conv7
shapes: torch.Size([48, 32]) torch.Size([48, 32])
max diff: 2.60770320892334e-08
mean diff: 5.162308802653115e-09


#### Norm5

In [ ]:
norm7.eval()
norm7.to(device)

BatchNorm(32, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)

In [ ]:
x1 = norm7(x1)
x2 = norm7(x2)
compare("norm7", x1, x2, batch_pair)

norm7
shapes: torch.Size([48, 32]) torch.Size([48, 32])
max diff: 2.60770320892334e-08
mean diff: 5.162308802653115e-09


In [ ]:
x1 = x1+x_sc1
x2 = x2+x_sc2

#### Pooling final

In [ ]:
x1_to_pool = x1.clone()
x2_to_pool = x2.clone()

In [ ]:
batch_single.pos.device

device(type='cuda', index=0)

In [ ]:
pool7.end.device

device(type='cpu')

In [ ]:
x1 = pool7(x1, pos=batch_single.pos[:, :2], batch=batch_single.batch)
x2 = pool7(x2, pos=batch_pair.pos[:, :2], batch=batch_pair.batch)

device end: tensor([120., 100.], device='cuda:0')
device end: tensor([120., 100.], device='cuda:0')


In [ ]:
print(x1.shape, x2.shape)


torch.Size([16, 32]) torch.Size([32, 32])


In [ ]:
x2_pool_g0 = x2[:x1.shape[0]]

In [ ]:
print("pool7 max diff:", (x1 - x2_pool_g0).abs().max().item())
print("pool7 mean diff:", (x1 - x2_pool_g0).abs().mean().item())

pool7 max diff: 5.960464477539063e-08
pool7 mean diff: 5.382389645092189e-09


In [ ]:
z1 = x1.reshape(batch_single.num_graphs, -1)
z2 = x2.reshape(batch_pair.num_graphs, -1)

print(z1.shape, z2.shape)
print("flatten max diff:", (z1 - z2[0:1]).abs().max().item())
print("flatten mean diff:", (z1 - z2[0:1]).abs().mean().item())

torch.Size([1, 512]) torch.Size([2, 512])
flatten max diff: 5.960464477539063e-08
flatten mean diff: 5.382389645092189e-09


In [ ]:
fc.eval()
fc.to(device)

Linear(in_features=512, out_features=2, bias=True)

In [ ]:
out1 = fc(z1)
out2 = fc(z2)

print("fc max diff:", (out1 - out2[0:1]).abs().max().item())
print("out1:", out1)
print("out2[0]:", out2[0:1])

fc max diff: 0.0
out1: tensor([[-0.0095,  0.0157]], device='cuda:0', grad_fn=<AddmmBackward0>)
out2[0]: tensor([[-0.0095,  0.0157]], device='cuda:0', grad_fn=<SliceBackward0>)


### Debug pooling

In [ ]:
from torch_geometric.nn.pool import max_pool_x, voxel_grid

In [ ]:
voxel_size = input_shape[:2] //4
voxel_size = voxel_size.to(device)

batch_single.pos.device


device(type='cuda', index=0)

In [ ]:
voxel_size.device

device(type='cuda', index=0)

In [ ]:
# size = 16
# if torch.is_tensor(size):
#     size = size.to(device=pos.device, dtype=pos.dtype)
# elif isinstance(size, (list,tuple)):
#     size = [float(v) for v in size]

In [ ]:
cluster_single= voxel_grid(batch_single.pos, batch=batch_single.batch, size=voxel_size, start=[0., 0.], end=input_shape[:2].to(device))

print("cluster min/max:", cluster_single.min(), cluster_single.max())
print("unique clusters:", len(torch.unique(cluster_single)))

cluster min/max: tensor(0, device='cuda:0') tensor(12, device='cuda:0')
unique clusters: 9


In [ ]:
cluster_single

tensor([ 0,  0,  0,  1,  1,  1,  2,  2,  0,  0,  0,  1,  1,  1,  2,  2,  0,  0,
         5,  6,  1,  1,  2,  7,  5,  5,  5,  6,  6,  6,  7,  7,  5,  5,  5,  6,
         6,  6,  7,  7, 10, 10, 10, 11, 11, 11, 12, 12], device='cuda:0')

In [ ]:
x2_to_pool.shape

torch.Size([131, 32])

In [ ]:
cluster_pair = voxel_grid(batch_pair.pos, batch=batch_pair.batch, size=voxel_size, start=[0., 0.], end=input_shape[:2].to(device))

print("cluster min/max:", cluster_pair.min(), cluster_pair.max())
print("unique clusters:", len(torch.unique(cluster_pair)))

cluster min/max: tensor(0, device='cuda:0') tensor(138, device='cuda:0')
unique clusters: 21


In [ ]:
cluster_pair[batch_pair.batch == 0]

tensor([ 0,  0,  0,  1,  1,  1,  2,  2,  0,  0,  0,  1,  1,  1,  2,  2,  0,  0,
         5,  6,  1,  1,  2,  7,  5,  5,  5,  6,  6,  6,  7,  7,  5,  5,  5,  6,
         6,  6,  7,  7, 10, 10, 10, 11, 11, 11, 12, 12], device='cuda:0')

In [ ]:
mask0 = batch_pair.batch == 0

c1 = cluster_single
c2 = cluster_pair[mask0]

print(c1.shape, c2.shape)
print("same cluster ids:", torch.equal(c1, c2))
print("max abs diff:", (c1 - c2).abs().max())
print("num different:", (c1 != c2).sum())

torch.Size([48]) torch.Size([48])
same cluster ids: True
max abs diff: tensor(0, device='cuda:0')
num different: tensor(0, device='cuda:0')


In [ ]:
c1 = cluster_single
c2 = cluster_pair[batch_pair.batch == 0]

print(c1)
print(c2)

for i in torch.unique(c1):
    print(
        i.item(),
        torch.where(c1 == i)[0],
        torch.where(c2 == c2[torch.where(c1 == i)[0][0]])[0]
    )

tensor([ 0,  0,  0,  1,  1,  1,  2,  2,  0,  0,  0,  1,  1,  1,  2,  2,  0,  0,
         5,  6,  1,  1,  2,  7,  5,  5,  5,  6,  6,  6,  7,  7,  5,  5,  5,  6,
         6,  6,  7,  7, 10, 10, 10, 11, 11, 11, 12, 12], device='cuda:0')
tensor([ 0,  0,  0,  1,  1,  1,  2,  2,  0,  0,  0,  1,  1,  1,  2,  2,  0,  0,
         5,  6,  1,  1,  2,  7,  5,  5,  5,  6,  6,  6,  7,  7,  5,  5,  5,  6,
         6,  6,  7,  7, 10, 10, 10, 11, 11, 11, 12, 12], device='cuda:0')
0 tensor([ 0,  1,  2,  8,  9, 10, 16, 17], device='cuda:0') tensor([ 0,  1,  2,  8,  9, 10, 16, 17], device='cuda:0')
1 tensor([ 3,  4,  5, 11, 12, 13, 20, 21], device='cuda:0') tensor([ 3,  4,  5, 11, 12, 13, 20, 21], device='cuda:0')
2 tensor([ 6,  7, 14, 15, 22], device='cuda:0') tensor([ 6,  7, 14, 15, 22], device='cuda:0')
5 tensor([18, 24, 25, 26, 32, 33, 34], device='cuda:0') tensor([18, 24, 25, 26, 32, 33, 34], device='cuda:0')
6 tensor([19, 27, 28, 29, 35, 36, 37], device='cuda:0') tensor([19, 27, 28, 29, 35, 36, 37],

In [ ]:
from torch_geometric.nn.pool.consecutive import consecutive_cluster


In [ ]:
from torch_geometric.nn.pool.consecutive import consecutive_cluster

c1, _ = consecutive_cluster(cluster_single)

mask = batch_pair.batch == 0
c2, _ = consecutive_cluster(cluster_pair[mask])

print(torch.equal(c1, c2))

True


In [ ]:
from torch_geometric.nn.pool import max_pool_x, voxel_grid

In [ ]:
x1, _ = max_pool_x(c1, x1_to_pool, batch_single.batch, size=16)

x2, _ = max_pool_x(
    c2,
    x2_to_pool[mask0],
    torch.zeros_like(batch_pair.batch[mask0]),
    size=16
)

print((x1 - x2).abs().max())

tensor(8.9407e-08, device='cuda:0', grad_fn=<MaxBackward1>)


##### Rectification

In [ ]:
cluster_single = voxel_grid(batch_single.pos, batch=batch_single.batch, size=size)
batch_single.batch

cluster_fixed = torch.empty_like(cluster_single)

max_index = 0
print(f'max_index: {max_index}')

for gid in torch.unique(batch_single.batch):
    mask = batch_single.batch == gid
    print(f'cluster n°{gid} : {cluster_single[mask]}')
    c_local, _ = consecutive_cluster(cluster_single[mask])
    # print(c_local)
    # assert c_local.max().item() < size, (
    #     c_local.max().item() + 1,
    #     size
    # )

    cluster_fixed[mask] = c_local + max_index
    # print( c_local.max().item())
    max_index +=  c_local.max().item() + 1
    print(f'max_index: {max_index}')

cluster_single = cluster_fixed

x1, _ = max_pool_x(cluster_single, x1_to_pool, batch_single.batch, size=size)

max_index: 0
cluster n°0 : tensor([ 0,  0,  1,  2,  2,  3,  3,  4,  0,  0,  1,  2,  2,  3,  3,  4,  5,  5,
         6,  7,  7,  8,  8,  9, 10,  5,  6,  7, 12,  8,  8,  9, 10, 10, 11, 11,
        12, 13, 13, 14, 15, 15, 16, 16, 17, 18, 18, 19], device='cuda:0')
max_index: 20


In [ ]:
cluster_single.unique()

tensor([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
        18, 19], device='cuda:0')

In [ ]:
cluster_pair = voxel_grid(batch_pair.pos, batch=batch_pair.batch, size=size)
print(len(cluster_pair.unique()))
# batch_single.batch

cluster_fixed = torch.empty_like(cluster_pair)

max_index = 0
print(f'max_index: {max_index}')

for gid in torch.unique(batch_pair.batch):
    mask = batch_pair.batch == gid
    print(f'cluster n°{gid} : {cluster_pair[mask]}')
    c_local, _ = consecutive_cluster(cluster_pair[mask])
    # print(c_local)
    # assert c_local.max().item() < size, (
    #     c_local.max().item() + 1,
    #     size
    # )

    cluster_fixed[mask] = c_local + max_index
    # print( c_local.max().item())
    max_index +=  c_local.max().item() + 1
    print(f'max_index: {max_index}')

cluster_pair = cluster_fixed
print(cluster_pair.unique)
x2, _ = max_pool_x(cluster_pair, x2_to_pool, batch_pair.batch, size=size)

48
max_index: 0
cluster n°0 : tensor([ 0,  0,  1,  2,  2,  3,  3,  4,  0,  0,  1,  2,  2,  3,  3,  4,  7,  7,
         8,  9,  9, 10, 10, 11, 14,  7,  8,  9, 16, 10, 10, 11, 14, 14, 15, 15,
        16, 17, 17, 18, 21, 21, 22, 22, 23, 24, 24, 25], device='cuda:0')
max_index: 20
cluster n°1 : tensor([28, 28, 29, 30, 30, 31, 31, 32, 33, 33, 34, 34, 28, 28, 29, 29, 30, 31,
        31, 32, 33, 33, 34, 35, 35, 36, 37, 37, 38, 38, 39, 40, 40, 41, 41, 35,
        35, 36, 36, 37, 45, 45, 46, 40, 40, 41, 48, 42, 42, 43, 43, 44, 45, 45,
        46, 47, 47, 48, 48, 49, 49, 50, 50, 51, 52, 52, 53, 54, 54, 55, 55, 49,
        49, 50, 51, 51, 52, 52, 53, 54, 54, 55, 55], device='cuda:0')
max_index: 48
<bound method Tensor.unique of tensor([ 0,  0,  1,  2,  2,  3,  3,  4,  0,  0,  1,  2,  2,  3,  3,  4,  5,  5,
         6,  7,  7,  8,  8,  9, 10,  5,  6,  7, 12,  8,  8,  9, 10, 10, 11, 11,
        12, 13, 13, 14, 15, 15, 16, 16, 17, 18, 18, 19, 20, 20, 21, 22, 22, 23,
        23, 24, 25, 25, 26, 26, 2

In [ ]:
z1 = x1.reshape(batch_single.num_graphs, -1)
z2 = x2.reshape(batch_pair.num_graphs, -1)

print(z1.shape, z2.shape)
print("flatten max diff:", (z1 - z2[0:1]).abs().max().item())
print("flatten mean diff:", (z1 - z2[0:1]).abs().mean().item())

torch.Size([1, 512]) torch.Size([2, 512])
flatten max diff: 8.940696716308594e-08
flatten mean diff: 1.6227204469032586e-08


In [ ]:
out1 = fc(z1)
out2 = fc(z2)

print("fc max diff:", (out1 - out2[0:1]).abs().max().item())
print("out1:", out1)
print("out2[0]:", out2[0:1])

fc max diff: 5.960464477539063e-08
out1: tensor([[-0.1480, -0.0309]], device='cuda:0', grad_fn=<AddmmBackward0>)
out2[0]: tensor([[-0.1480, -0.0309]], device='cuda:0', grad_fn=<SliceBackward0>)


# Code de test

In [ ]:
num_params = sum(p.numel() for p in model.parameters())
print(num_params)

30354


In [ ]:
import yaml

In [ ]:
def load_config(config_path: str) -> dict:
    with open(config_path, "r", encoding="utf-8") as f:
        return yaml.safe_load(f)

In [ ]:
cfg = load_config("../checkpoints/aegnn/recognition/ncars/v2/config.yaml")

In [ ]:
import torch

x = torch.tensor([[1,2,3],[4,5,6],[7,8,9]])
x1 = x/torch.max(x)

print(x1)

tensor([[0.1111, 0.2222, 0.3333],
        [0.4444, 0.5556, 0.6667],
        [0.7778, 0.8889, 1.0000]])


In [ ]:
root = "/home/ibrahim/event_graph_aegnn/test_codes/data/ncars/processed_adaptatif/test"

missing_y = []
empty_y = []
nan_y = []
bad_labels = []
i = 0
for seq_name in os.listdir(root): 
    i +=1
    seq_path = os.path.join(root, seq_name)

    if not os.path.isfile(seq_path):
        continue

    data = torch.load(seq_path, weights_only=False)

    # y absent
    if not hasattr(data, "y"):
        missing_y.append(seq_name)
        continue

    # y vide
    if data.y.numel() == 0:
        empty_y.append(seq_name)
        continue

    # y contient des NaN
    if torch.isnan(data.y.float()).any():
        nan_y.append(seq_name)
        continue

    # labels différents de 0 ou 1
    unique_labels = torch.unique(data.y)

    if not torch.all((unique_labels == 0) | (unique_labels == 1)):
        bad_labels.append((seq_name, unique_labels.tolist()))
print("Number of sequences: ", i)
print("Missing y:", len(missing_y))
print("Empty y:", len(empty_y))
print("NaN y:", len(nan_y))
print("Bad labels:", len(bad_labels))

Number of sequences:  8606
Missing y: 0
Empty y: 0
NaN y: 0
Bad labels: 0


In [10]:
import torch
# quick sanity check — run this on one of your loaded graphs
data = torch.load("/home/ibrahim/event_graph_aegnn/test_codes/data/ncars/beta_processed/aegnn_raduis_preprocessing/xytp_beta_default/ncars/processed/training/sequence_1", weights_only=False, map_location="cpu")
src, dst = data.edge_index
print("Are edges symmetric?", (data.edge_index.shape[1] % 2 == 0))
print("Sample src:", src[:5])
print("Sample dst:", dst[:5])
# check if for every (i,j) there is a (j,i)
edges = set(zip(src.tolist(), dst.tolist()))
symmetric = all((j, i) in edges for i, j in edges)
print("Symmetric graph:", symmetric)

Are edges symmetric? False
Sample src: tensor([  6,  18,  34,  68, 115])
Sample dst: tensor([0, 0, 0, 0, 0])
Symmetric graph: False


In [20]:
N = 1
x = torch.randn(N, 10)
x_reduced = x.reshape(N, 10//2, 2).mean(dim=2)  # [N, 16]
print(x)
print(x_reduced)

tensor([[ 0.3862,  0.4600,  0.6460, -1.8539, -0.0147, -1.4982, -1.6853,  1.7516,
         -0.1599,  0.1207]])
tensor([[ 0.4231, -0.6039, -0.7564,  0.0331, -0.0196]])


In [ ]:
dict = {"data_params": {"method":1, "n":2}}


SyntaxError: invalid syntax (3329426196.py, line 2)